---
title: "Archive"
author: "Carolina Lobato"
date: "28/02/2023"
output: html_document
---

## Plot significance

**Work in progress**: Add to plot

```{r, echo = FALSE}
sig <- compare_means(Shannon ~ cultivar,
                     data = alpha,
                     p.adjust.method = "bonferroni")
sig_adj <- select(filter(sig, p.adj < 0.05), 
                 c(group1, group2, p.adj, p.signif))
class(sig_adj)
sample_pair <- data.frame(sig_adj$group1, sig_adj$group2)

my_comparisons <- list()
for (idx in 1:nrow(sample_pair)) {
  g1 <- sample_pair[idx, "sig_adj.group1"]
  g2 <- sample_pair[idx, "sig_adj.group2"]
  my_comparisons[[idx]] <- c(g1, g2)
}
str(my_comparisons) # print

symnum.args <- list(
  cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
  symbols = c("****", "***", "**", "*", "ns")
  ) # add significance comparisons

cld <- multcompLetters4()
```

## Plot Alpha-diversity displaying traits

```{r, echo = FALSE}
# bh_set <- subset_samples(ps_raref, species.1 != "" & inflorescence == "dioecious"); bh_set # select samples
# g.rate2 <- sample_data(ps_raref)$g.rate2
# g.rate2[g.rate2 == ""] <- NA
# sample_data(bh_raref)$g.rate2 <- factor(g.rate2, levels = c("below-0.85", "above-0.85"))

mycolors <- colorRampPalette(brewer.pal(4, "Set1"))(4) # define colors

ggplot(alpha, aes(x = reorder(cultivar, Shannon), y = Shannon)) +
  geom_point(
    aes(color = cultivar, shape = chemotype2),
    size = 1.5 ,
    alpha = 1
  ) +
  geom_boxplot(alpha = 0.1, show.legend = FALSE) +
  theme_bw() +
  theme(
    axis.text.x = (element_text(angle = 45, hjust = 1)),
    legend.position = "bottom",
    legend.box = "vertical",
    legend.direction = "horizontal",
    strip.background = element_blank(),
    strip.text = element_blank()
  ) +
  scale_shape_manual(values = c(19, 1, 17)) +
  scale_color_manual(values = mycolors, na.translate = TRUE) +
  labs(color = "Application", shape = "Chemotype") +
  xlab("") +
  ylab("Shannon Index")
# coord_flip() +
# stat_compare_means(
#   method = "wilcox.test",
#   comparisons = as.list(my_comparisons),
#   symnum.args = symnum.args,
#   hide.ns = TRUE,
#   na.rm = T) +
# facet_wrap(~ c("Observed", "Shannon"), nrow = 2) #, vjust = 0.5, hjust = 0))

# ggsave(
#  "outputs/alpha-div/alpha-div-traits.png",
#  width = 20,
#  height = 18,
#  units = "cm",
#  scale = 1,
#  dpi = 320
#)
```

## Microbial Classification

> For core/flex, abundance-occupancy curves and spec/gen

***Code Shade lab: <https://github.com/ShadeLab/PAPER_Shade_CurrOpinMicro/blob/master/script/Core_prioritizing_script.R>***

```{r, echo = FALSE}
load("outputs/r/204_setup/bh_raref.RData")
load("outputs/r/spec-gen/spec-gen-1000-cultivar.RData")

asv.raref <- otu_table(bh_raref) 
taxo.raref <- tax_table(bh_raref) 
meta.raref <- meta(bh_raref) 

# presence-absence data
ASV_PA <- 1 * ((asv.raref > 0) == 1)

# occupancy calculation
ASV_prev <- rowSums(ASV_PA) / ncol(ASV_PA)

# relative abundance
ASV_rel <- apply(vegan::decostand(asv.raref, method = "total", MARGIN = 2), 1, mean)

# combining occupancy, relative abundance and spec/gen classification of each ASV in a table
ASV_prev_rel_gs <- rownames_to_column(as.data.frame(cbind(ASV_prev, ASV_rel)), "ASV") # , res4

## merging with metadata
ASV_prev_rel_gs_taxo <- rownames_to_column(as.data.frame(taxo.raref), "ASV") 
ASV_prev_rel_gs_taxo <- merge(ASV_prev_rel_gs_taxo, ASV_prev_rel_gs, by = "ASV")

ASV1 <- rownames_to_column(as.data.frame(t(asv.raref)), "replicates")
ASV_use1 <- merge(meta.raref, ASV1, by = "replicates")

# Keep only the one feature column as metadata for collapsing table
ASV_use1_1 <- ASV_use1[ -c(1, 3:12) ]

ASV_cultivar = ASV_use1_1 %>% group_by(cultivar) %>% summarise_each(list(sum))

# Make table as presence absence of ASV for each feature
ASV_cultivar_PA <- 1 * ((ASV_cultivar > 0) == 1)

## Merging ASV prevalence, RA and taxonomy with the number of varieties where each ASV was observed
ASV_obs_cultivar <- data.frame(colSums(ASV_cultivar_PA)) %>%
  na.omit() %>% rownames_to_column("ASV")
names(ASV_obs_cultivar)[names(ASV_obs_cultivar) == "colSums.ASV_cultivar_PA."] <- 'Obs'

ASV_prev_rel_gs_taxo_obs_cultivar <- merge(ASV_prev_rel_gs_taxo, ASV_obs_cultivar, by = "ASV")

save(ASV_prev_rel_gs_taxo_obs_cultivar, file = "outputs/r/303_bac_str/prev-rel-gs-taxo-obs-cultivar-raref.RData")
```

## Heatmap Study - jmdf

```{r}
## Study different ways of making the heatmap the core abundances
## Different normalization/transformations and different distances.

## TODO: calculate dist matrix with: vegdist(data_otu_filt_rar, method = "bray")
## TODO: use feature importance (SHAP values) from random-forest classifier to
## cluster ASVs

methods <- c(
  "total",
  "max",
  "frequency",
  "normalize",
  "range","rrank",
  "standardize",
  "hellinger",
  "log",
  #"alr",
  "clr",
  "rclr",
  "chi.square"
  )

clustering_methods <- c(
  "single", "complete", "average", "ward.D", "ward.D2", "centroid", "mcquitty"
)

## Core ASVs
coreASVs <- read.csv("data/core/coreASV.csv")[[1]]

## ASV-by-Cultivar
asv.by.cultivar <- read.csv(
  "data/phyloseq/asv-by-cultivar.csv",
  row.names = "ASV") %>%
  filter(row.names(.) %in% coreASVs)

## Filter ASV names
row.names(asv.by.cultivar) <- sub(
  ":[a-z0-9]*:[a-z_]{0,3}",
  " | ",
  row.names(asv.by.cultivar)
  )

topK <- 100
for (cmethod in clustering_methods) {
  for (method in methods) {
    print(paste0("Heatmap for ", method, " and ", cmethod))
    
    ## Standardize
    if (method == "clr") {
      std.counts <- decostand(asv.by.cultivar, method,  pseudocount = 1) 
    } else {
      std.counts <- decostand(asv.by.cultivar, method) 
    }
      
    sort.index <- order(rowSums(std.counts), decreasing=TRUE)
    std.counts.topK <- std.counts[sort.index[1:topK],]
    
    ## Heatmap
    color <- colorRampPalette(c("lightgreen", "orange", "red"))(256)
    pheatmap(
      std.counts.topK, 
      color=color,
      clustering_method="complete",
      clustering_distance_rows="manhattan",  # 2x Bray-curtis
      clustering_distance_cols="manhattan",  # 2x Bray-curtis
      scale="none",
      treeheight_row=100,
      fontsize_col=9,
      fontsize_row=10,
      filename=paste0("output/heatmaps/", method, "-", cmethod ,"-core.png"),
      angle_col=45,
      height=12,
      width=12,
    )
  }
}
```

## Alpha-diversity

From Marie Simonin

```{r alpha diveristy}
Div16S <- read.table("phyloseq-obj/metadata.csv", header = TRUE, check.names = FALSE, sep="\t")

head(Div16S)
dim(Div16S)

Diversity_stat <- summarySE(Div16S, measurevar="observed_otus", groupvars=c("Plant"), na.rm = TRUE)

# Calculate median by plant
Div16S <- Div16S %>% group_by(Plant) %>%mutate(median_observed_otus_plant = median(observed_otus))

# Calculate stats by family
stat_family = Div16S %>%
  group_by(Family) %>%
  summarise_each(funs(max, min, mean, median, sd), observed_otus)
stat_family

Div16S$Plant <- ordered(Div16S$Plant,
                      levels = c("Cauliflower","Tomato","Broccoli","Tobacco","Cabbage","Eyebright","Turnip","Willow Gentian","Lolium arundinacea","Phelipanche ramosa", "Garden rocket","Great Masterwort","Pincushion Flower","Festuca rubra","Setaria pumila","Wheat","Setaria viridis","Rapeseed","Capsella bursa-pastoris","Grass of Parnassus", "Lolium perenne","Cardamine hirsuta","Melon","Brassica nigra","Erophila verna ", "Alliaria petiolata","Arabidopsis thaliana","Pea", "Heliosperma alpestre","Rorippa sylvestris","Siberian wildrye", "Carrot","Barbarea vulgaris ","Rice","Rhinanthus glacialis","Medicago truncatula","Oat","Dahurian wildrye", "Sinapis arvensis","Radish","Bean","Hairy vetch", "Sunflower","Chiltern Gentian","Berteroa incana", "Alfalfa", "Oak"))

color = c("#5892ae","#ff2600","#aea9ca","#ff9b7c","#811770","#00bb9f","#c5d6d6","#15cd7e","#3c884c","#08ffda","#67bde8","#f0810F","#e09aa5","#2E4600","#c5cd77","#fdf351","#edf050","#008bc4","#6b66a3","#d0425d","#72ae4f","#1b2b7d","#FA6775","#a96699","#5e87c5","#4a1777","#835e9f","#68104d","#ffceda","#3e8a89","#9ff76e","#ee8332","#baa5c8","#b2d24f","#8eddad","#b554a6","#66ff00","#2fbd03","#9abdbb","#4cb5f5","#e03581","#e035ac","#fec767","#00a7b5","#ceaac4","#e035d7","#000000")

p4 = ggplot(data = Div16S, aes(x = observed_otus, y = Study_ID, color = Plant, shape = Seed_fraction)) +
  geom_jitter(alpha = 0.8) +
  xlab("Observed ASV richness") +
  ylab("Studies") +
  theme_gray() +
  theme(axis.title = element_text(color = "black", size = 9, face = "bold")) +
  theme(axis.text = element_text(color = "black", size = 7, face = "bold")) +
  facet_grid(Plant~., scales = "free", space = "free") +
  theme(legend.text = element_text(color = "black", size = 8, face = "bold")) +
  theme(strip.text.y = element_text(size = 8, angle = 0, face = "bold", margin = margin(b = 2, t = 2))) +
  scale_shape_manual(values = c(3, 1, 2, 16)) +
  geom_vline(aes(xintercept = median_observed_otus_plant, group = Plant, colour = Plant)) +
  labs(shape = "Gene Region", color = "Plant Species") +
  theme(legend.title = element_text(color = "black", size = 12, face = "bold")) +
  scale_x_log10() +
  scale_color_manual(values = color) +
  theme(panel.background = element_rect(fill = "#eeeeee",colour = "#eeeeee",size = 0.9, linetype = "solid"), panel.grid.major = element_line(size = 0.5, linetype = 'solid',colour = "white"), panel.grid.minor = element_line(size = 0.25, linetype = 'solid', colour = "#eeeeee")) + theme(strip.background = element_rect(fill = "#d9dbdb")) +
  ggtitle("16S rRNA gene - Bacteria & Archaea") +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 10)) +
  theme(panel.spacing.y = unit(0.2, "lines")) +
  guides(col = guide_legend(ncol = 4))
p4


```

## Other

### Agglomerate taxa

```{r, echo = FALSE}
# df <- psmelt(bh_cannabis)
# kable(head(df))

# Working with different taxonomic levels
bh_spe <- phyloseq::tax_glom(bh_cannabis,
                             "Species",
                             NArm = TRUE) # NArm removes anything that is unclassified
bh_spe # 163 taxa
# taxa_names(bh_spe) <- tax_table(bh_spe)[,"Species"] # Assign taxa names
unique(tax_table(bh_spe)[, "Species"]) # 156 taxa

ps_gen <- phyloseq::tax_glom(ps_cannabis,
                             "Genus",
                             NArm = TRUE) # NArm removes anything that is unclassified)
                             ps_gen # 676 taxa
                             
# taxa_names(ps_gen) <- tax_table(ps_gen)[,"Genus"] # Assign taxa names
unique(tax_table(ps_gen)[, "Genus"]) # 675 taxa
                             
ps_fam <- phyloseq::tax_glom(ps_cannabis,
                             "Family",
                             NArm = TRUE) # NArm removes anything that is unclassified
ps_fam # 364 taxa
# taxa_names(ps_fam) <- tax_table(ps_fam)[,"Family"] # Assign taxa names
unique(tax_table(ps_fam)[, "Family"]) # 364 taxa
                             
ps_ord <- phyloseq::tax_glom(ps_cannabis,
                             "Order",
                             NArm = TRUE) # NArm removes anything that is unclassified
ps_ord # 223
# taxa_names(ps_ord) <- tax_table(ps_ord)[,"Order"] # Assign taxa names
unique(tax_table(ps_fam)[, "Order"]) # 220 taxa
                             
ps_cla <- phyloseq::tax_glom(ps_cannabis,
                             "Class",
                             NArm = TRUE) # NArm removes anything that is unclassified
ps_cla # 96 taxa
# taxa_names(ps_cla) <- tax_table(ps_cla)[,"Class"] # Assign taxa names
unique(tax_table(ps_cla)[, "Class"]) # 96 taxa
                             
ps_phy <- phyloseq::tax_glom(ps_cannabis,
                             "Phylum",
                             NArm = TRUE) # NArm removes anything that is unclassified
ps_phy # 37 taxa
taxa_names(ps_phy) <-
tax_table(ps_phy)[, "Phylum"] # Assign taxa names
unique(tax_table(ps_phy)[, "Phylum"]) # 37 taxa
```

### Heatmap

```{r, echo = FALSE}
bh_c75 <- subset_samples(bh_raref, harvest.days != "")
bh_c75 <- phyloseq::merge_samples(bh_c75, "harvest.days")
bh_c75 <- microbiome::transform(bh_c75, "compositional")
bh_c75 <- aggregate_rare(
  bh_c75,
  level = "Genus",
  NArm = TRUE,
  detection = .1 / 100,
  prevalence = 75 / 100
)

days <- factor(sample_data(bh_c75)$harvest.days,
               levels = sort(unique(sample_data(bh_c75)$harvest.days)))

heat_core <- (
  bh_c75 %>%
    microbiome::plot_composition(
      sample.sort = "neatmap",
      otu.sort = "neatmap",
      plot.type = "heatmap"
    ) +
    ylab("Germination rate") +
    theme(
      axis.text.y = element_text(size = 8, hjust = 1),
      axis.ticks.y = element_blank(),
      axis.text.x = element_text(
        size = 8,
        hjust = 1,
        angle = 30
      ),
      legend.text = element_text(size = 8)
    ) +
    scale_fill_gradient(
      limits = c(0, 1),
      low = "white",
      high = "#9c0404",
      name = ""
    )
)
print(heat_core)
```

### Relative Abundance

```{r, echo = FALSE}
#bh_rel1 <- microbiome::transform(bh_cannabis, "compositional"); otu_table(bh_rel1) # RA of each ASV per sample
#bh_rel2 <- decostand(otu_table(bh_cannabis), method = "total", MARGIN = 2); bh_rel2 # RA of each ASV per sample
#bh_rel3 <- relative_abundance(bh_cannabis); bh_rel3 # RA of each ASV in the whole dataset
#bh_rel4 <- apply(decostand(otu_table(bh_cannabis), method = "total", MARGIN = 2), 1, mean); bh_rel4  # mean RA of each ASV in the whole dataset

bh_taxa <- subset_taxa(bh_cannabis, Genus == "Bacillus"); bh_taxa
a <- abundances(bh_taxa) # count reads
sum(a)
nsamples(bh_taxa)
```
