---
title: "metabarcoding-analysis-pipeline"
author: "Carolina Lobato"
date: "6/2/2022"
output: html_document
---

```{r setup, include = FALSE}

knitr::opts_knit$set(root.dir = "/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB-all/rstudio")

getwd()

# Useful links:
# https://benjjneb.github.io/dada2/tutorial.html
# https://benjjneb.github.io/decontam/vignettes/decontam_intro.html
# https://www.nicholas-ollberding.com/post/introduction-to-phyloseq/
# https://web.stanford.edu/class/bios221/labs/phyloseq/lab_phyloseq.html
# https://genoweb.toulouse.inra.fr/~formation/15_FROGS/8-February2017/FROGS_phyloseq_02_2017.pdf
# https://microbiome.github.io/tutorials/Core.html
# https://userweb.eng.gla.ac.uk/umer.ijaz/projects/microbiomeSeq_Tutorial.html#beta_div
```

```{r loadlib, echo = FALSE}

# if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")

# if (!requireNamespace("devtools", quietly = TRUE))
#      install.packages("devtools")

# BiocManager::install(c("decontam", "phyloseq", "microbiome", "edgeR", "rnaseqGene",  "metagenomeSeq", "multtest")) 

# install.packages(c("remotes", "tidyverse", "vegan", "genefilter", "DESeq2", "psych", "phangorn"))

library(devtools)
# BiocManager::install("dada2", version = "3.15")
library(dada2)
library(tidyverse)
library(RColorBrewer)
library(decontam)
library(phyloseq)
library(biomformat)
library(vegan)
library(genefilter)
library(psych)
library(DESeq2)
library(reshape2)
library(microbiome)
# install_github("umerijaz/microbiomeSeq")
remotes::install_github("microbiome/microbiome")
# devtools::install_github("microsud/microbiomeutilities")
library(microbiomeutilities)
library(knitr)
# devtools::install_github("kassambara/ggpubr")
library(ggpubr)
# remotes::install_github("vmikk/metagMisc")
library(metagMisc)
# devtools::install_github("gauravsk/ranacapa")
library(ranacapa)
library(Biostrings)
library(metagenomeSeq)
# install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)
# BiocManager::install("treeio")
library(treeio)
# remotes::install_github("mikemc/speedyseq")
library(speedyseq)
library(ggsignif)
remotes::install_github("cpauvert/psadd")
```

```{r import, echo = FALSE}

setwd("/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB-all/rstudio")
getwd()

### Troubleshooting: 
  # make sure the column names correspond between files (have the same order)
  # if there are any blank cells use the fill option
  # don't use tax_table as a name, as it is a phyloseq function and you will overwrite it
  # check if the metadata is properly distributed into columns inside r
  # sample names should have _ instead of -

## Import metadata file
metadata <- read.table("phyloseq-obj/metadata.csv", 
                       sep = "\t", 
                       header = TRUE, 
                       fill = TRUE, 
                       row.names = "sampleid"
                       )
metadata$replicates <- row.names(metadata)
# metadata$batch <- ifelse(metadata$type == "sample1" | metadata$type == "poscontrol1" | metadata$type == "negcontrol1", "pool1", "pool2")
# metadata$is.neg <- metadata$type == "negcontrol1" | metadata$type == "negcontrol2"
metadata <- sample_data(metadata)
attributes(metadata)
summary(metadata)  
  
## Import filtered feature table file
asv_table <- read.table("phyloseq-obj/exp-filter-table.tsv", 
                        header = TRUE, 
                        row.names = "OTUID"
                        )
feature_table <- otu_table(asv_table, 
                           taxa_are_rows = TRUE
                           )
head(feature_table)
summary(feature_table)

## Import taxonomy file
taxonomy_table <- read.table("phyloseq-obj/tax-vsearch.tsv", 
                             header = TRUE, 
                             fill = TRUE, 
                             row.names = "OTUID"
                             )
taxonomy_table <- as.matrix(taxonomy_table)
taxonomy_table <- tax_table(taxonomy_table)
head(taxonomy_table)
summary(taxonomy_table)

## Import filtered reference sequences file
rs_table <- readDNAStringSet("phyloseq-obj/dna-sequences.fasta", 
                             format = "fasta", 
                             nrec = -1L, 
                             skip = 0L, 
                             seek.first.rec = FALSE, 
                             use.names = TRUE, 
                             with.qualities = FALSE
                             )
rs_table
attributes(rs_table)

## Import rooted tree file
tree <- read_tree("phyloseq-obj/tree.nwk")
tree
attributes(tree)

## Convert to phyloseq object
ps <- merge_phyloseq(feature_table, 
                     taxonomy_table, 
                     metadata, 
                     rs_table, 
                     tree
                     ) # includes 3 pools
ps # 6695 taxa and 467 samples

sample_variables(ps)
sort(sample_sums(ps))
summary(sample_sums(ps)) # mean reads/sample = 85884
hist(sample_sums(ps), 
     main = "Histogram: Read Counts", 
     xlab = "Total Reads", 
     border = "black", 
     col = "orange", 
     las = 1, 
     breaks = 12
     )

rank_names(ps)
head(tax_table(ps))
head(taxa_sums(ps))
```

```{r unfiltered samples stats, echo = FALSE}

getwd()

# Check reads and ASVs before filtering mitochondria and chloroplasts and remove controls (use the unfiltered Rep Seqs and ASV-table for this phyloseq object)

## Import non-filtered feature table file
asv_table_unfil <- read.table("phyloseq-obj/exp-filter-table.tsv", 
                              header = TRUE, 
                              row.names = "OTUID"
                              )
feature_table_unfil <- otu_table(asv_table_unfil, 
                                 taxa_are_rows = TRUE
                                 )
head(feature_table_unfil)
summary(feature_table_unfil)

## Import non-filtered reference sequences file
rs_table_unfil <- readDNAStringSet("phyloseq-obj/dna-sequences.fasta",
                                   format = "fasta", 
                                   nrec = -1L, 
                                   skip = 0L, 
                                   seek.first.rec = FALSE, 
                                   use.names = TRUE, 
                                   with.qualities = FALSE
                                   )
head(rs_table_unfil)
attributes(rs_table_unfil)

## Convert to phyloseq object
ps_unfil <- merge_phyloseq(feature_table_unfil, 
                           taxonomy_table, 
                           metadata, 
                           rs_table_unfil
                           ) # includes both pools
ps_unfil # 4842 taxa and 309 samples

sort(sample_sums(ps_unfil))
summary(sample_sums(ps_unfil)) # mean reads/sample = 114997

ps_unfil <- subset_samples(ps_unfil, 
                           cultivar != ""
                           ) # exclude controls
ps_unfil # 4842 taxa and 303 samples
summary(sample_sums(ps_unfil)) # mean reads/sample = 115085
```

```{r subseting, echo = FALSE}

# The subset_taxa and subset_samples functions are for sub-setting based on auxiliary data contained in the Taxonomy Table or Sample Data components, respectively.
# For example, the following code will assign to ps_sub the subset of the ps dataset that are part of the d__Bacteria domain. 
# The prune_taxa function will remove the ASVs with 0 total reads throughout all samples that resulted from subset_taxa function.

## Subset bacteria
ps # 6695 taxa and 467 samples
ps_sub <- subset_taxa(ps, 
                      Domain == "d__Bacteria"
                      )
ps_sub <- prune_taxa(taxa_sums(ps_sub) > 0, ps_sub)
ps_sub # 5990 taxa and 467 samples

## Subset samples based on minimum number of reads
ps_sub2 <- prune_samples(sample_sums(ps_sub) > 1000, ps_sub)
ps_sub2

summarize_phyloseq(ps_sub2) # Average number of reads = 85642, Total number of reads = 39995034, Singletons = 0%
ss <- sample_sums(ps_sub2)
ss <- as.matrix(ss)
colSums(ss) # total reads = 26808205
summary(ss) # mean reads/sample = 85642

metadata$total_reads <- sample_sums(ps_sub2)
```

```{r decontam inspect, echo = FALSE}

df <- as.data.frame(sample_data(ps_sub)) # put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(ps_sub)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

df.sub <- subset(df, 
                 batch == "pool1"
                 )
# df.sub <- subset(df, batch == "pool2")
ggplot(data = df.sub, aes(x = Index, 
                        y = LibrarySize, 
                        color = type)
       ) + 
  geom_point()
```

```{r decontam frequency, echo = FALSE}

contamdf.freq <- isContaminant(ps_sub, 
                               method = "frequency", 
                               conc = "quant.reading", 
                               batch = "batch"
                               )
table(contamdf.freq$contaminant)
head(which(contamdf.freq$contaminant))

contamdf.freq.sub <- subset(contamdf.freq, 
                            contaminant == TRUE
                            )
row.names(contamdf.freq.sub)

plot_frequency(
  ps_sub,
  taxa_names(ps_sub)[c(694, 1006, 2069, 3674)],
  conc = "quant_reading"
  ) + 
  xlab("DNA Concentration (Nanodrop)")

ps_sub
ps_freq <- prune_taxa(!contamdf.freq$contaminant,
                      ps_sub
                      )
ps_freq

# set.seed(100)
# plot_frequency(ps_sub, taxa_names(ps_sub)[sample(which(contamdf.freq$contaminant),3)],  conc="quant_reading") + xlab("DNA Concentration (Nanodrop)")
```

```{r decontam prevalence, echo = FALSE}

contamdf.prev <- isContaminant(ps_sub, 
                               method = "prevalence",
                               neg = "is.neg",
                               batch ="batch",
                               threshold = 0.5
                               ) # a more strict approach
table(contamdf.prev$contaminant)
head(which(contamdf.prev$contaminant))

# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(ps_sub,
                                 function(abund) 1*(abund > 0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$is.neg,
                           ps.pa)
ps.pa.pos <- prune_samples(!sample_data(ps.pa)$is.neg,
                           ps.pa)

# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos = taxa_sums(ps.pa.pos),
                    pa.neg = taxa_sums(ps.pa.neg),
                    contaminant = contamdf.prev$contaminant)

ggplot(data = df.pa,
       aes(x = pa.neg,
           y = pa.pos,
           color = contaminant)) + 
  geom_point() +
  xlab("Prevalence (Negative Controls)") +
  ylab("Prevalence (True Samples)")

ps_prev <- prune_taxa(!contamdf.prev$contaminant,
                      ps_sub)
```

```{r decontam combined, echo = FALSE}

?isContaminant

contamdf.comb <- isContaminant(ps_sub, 
                               method = "combined", 
                               conc = "quant.reading",
                               batch.combine = "minimum",
                               neg = "is.neg",
                               batch = "batch",
                               threshold = 0.5
                               ) # a more strict approach
table(contamdf.comb$contaminant)
head(which(contamdf.comb$contaminant))

ps_sub 
ps_comb <- prune_taxa(!contamdf.comb$contaminant,
                      ps_sub
                      )
ps_comb 

sum_comb <- sample_sums(ps_comb)
sum_comb <- as.matrix(sum_comb)
colSums(sum_comb)  
summary(sum_comb) # consider median value for rarefaction
```

```{r subsetting samples and cleaning, echo = FALSE}

ps_cannabis <- subset_samples(ps_sub,
                              cultivar != ""
                              )
ps_cannabis <- prune_taxa(taxa_sums(ps_cannabis) > 0,
                          ps_cannabis
                          )
ps_cannabis # 5876 taxa and 458 samples

sort(sample_sums(ps_cannabis))
summary(sample_sums(ps_cannabis)) # mean reads/sample = 85040
summarize_phyloseq(ps_cannabis) # Average number of reads = 85040, Total number of reads = 38948469, Singletons = 0%

head(tax_table(ps_cannabis))

tax_table(ps_cannabis)[tax_table(ps_cannabis) == "d__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "p__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "c__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "o__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "f__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "f__Unknown_Family"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "g__uncultured"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "s__uncultured_bacterium"] <- ""
tax_table(ps_cannabis)[tax_table(ps_cannabis) == "s__metagenome"] <- ""
# taxa_names(ps_cannabis) <- gsub(" ", ".", taxa_names(ps_cannabis)) # spaces between names become _

head(tax_table(ps_cannabis))
summarize_phyloseq(ps_cannabis) # Average number of reads = 85040

otu_tab <- microbiome::abundances(ps_cannabis); otu_tab[1:5,1:5]
tax_tab <- phyloseq::tax_table(ps_cannabis); tax_tab[1:5,1:5]
taxa_names(ps_cannabis) <- gsub(taxa_names(ps_cannabis),
                                pattern = "d__", replacement = ""
                                )  
tax_table(ps_cannabis)[, colnames(tax_table(ps_cannabis))] <- gsub(tax_table(ps_cannabis)[, colnames(tax_table(ps_cannabis))],
                                                                   pattern = "[a-z]__",
                                                                   replacement = ""
                                                                   )
phyloseq::tax_table(ps_cannabis)[1:7,1:7]
```

```{r check samples in metadata, echo = FALSE}

ps_entry <- subset_samples(ps_cannabis,
                           inflorescence == "dioecious"
                           ); ps_entry # select samples
ps_mer <- merge_samples(ps_entry, "cultivar"); ps_mer
sample_names(ps_mer)
nsamples(ps_mer)

# total cultivars: n = 46

# monoecious: n = 9; "Beniko", "Fasamo", "Fedora17", "Felina32", "Ferimon", "Fibrimon", "Futura75", "Orion33", "Santhica27"
# dioecious: n = 26; "Arjans-Haze3", "Beniko", "Bredemann", "Carmagnola", "Elleta-Campana", "Fibridia", "Finola", "HavelÃ¤nder", "Hindu-Kush", "Kompolti", "Krasnoyarskaya", "Lovrin110", "Purini", "Ramo", "RV-Italy-84", "RV-Korea-87", "Schurig", "SilvanaxManitoba-Poison", "The-Church-CBD"

# low-cannabinoid: n = 28
# high-cannabinoid: n = 9; "Arjans-Haze3", "Hindu-Kush", "SilvanaxManitoba-Poison", "The-Church-CBD", "Afghani 1", "Master Kush", "Mexican Sativa", "Ruderalis Indica", "Skunk 1"

# sativa: n = 33
# chinensis: n = 1; "RV Korea 87"
# afghanica: n = 3; "Afghani 1", "Hindu Kush", "Master Kush"
# spontanea: n = 2; "Finola", "RV Georgia"
# gigantea: n = 1 "NA France"
# afghanica/spontanea: n = 1; "Ruderalis Indica"
# afghanica/indica: n = 4; "Arjans Haze 3", "Mexican Sativa", "Skunk 1", "The Church CBD"

# before-10: n = 46
# after-10: n = 7 ; "Arjans-Haze3", "BG Romania", "Bredemann", "Hindu-Kush", "Mexican Sativa", "NA France", "NA Syria"   

# above-0.85: n = 11; "Afghani 1", "Arjans-Haze3", "Elleta-Campana", "Fibrimon", "Master Kush", "Mexican Sativa", "Ruderalis Indica", "RV-Italy-88", "SilvanaxManitoba-Poison", "Skunk 1"  "The-Church-CBD"  
# below-0.85: n = 23

# early: n = 6; "Beniko", "Fasamo", "Fedora17", "Ferimon", "Finola", "Purini"    
# medium: n = 3; "Felina32", "Orion33", "Santhica27"
# late: n = 6; "Carmagnola", "Elleta-Campana", "Futura75", "Kompolti", "Krasnoyarskaya", "Lovrin110" 

# 1-5min: n = 27
# 1-7min: n = 19
# 1-10min: n = 8; "Bredemann", "Carmagnola", "Fasamo", "Lovrin110", "Ramo", "Santhica27", "Schurig"
# 1-15min: n = 3; "Carmagnola", "RV-Korea-87", "Santhica27"

# Europe/dioecious: n = 14
# Rest/dioecious: n = 12; "Afghani 1", "Arjans Haze 3", "Finola", "Hindu Kush",          "Master Kush", "Mexican Sativa", "Ruderalis Indica", "RV Georgia", "RV Korea 87", "Silvana x Manitoba Poison", "Skunk 1", "The Church CBD"    
```

```{r export qiime2 and microbiome analyst, echo = FALSE}

## Convert and export the phyloseq data into individual files

getwd()

# ASV table as tsv
asv_table_decontam = as(otu_table(ps_sub),
                        "matrix"
                        )
write.table(asv_table_decontam,
            "asv_decontam_r.tsv",
            sep = "\t",
            quote = FALSE
            )
# better export this format and then convert to biom as needs some modifications to be again compatible with qiime2 (check the R-imported file)

# ASV table as biom
asv_table_decontam <- otu_table(ps_sub)
asv_table_decontam <- make_biom(data = asv_table_decontam)
write_biom(asv_table_decontam,
           "asv_decontam_r.biom"
           )

# Taxonomy table as tsv
taxonomy_table_decontam = as(tax_table(ps_sub),
                             "matrix"
                             )
taxonomy_table_decontam = as.data.frame(taxonomy_table_decontam)
write.table(taxonomy_table_decontam,
            "taxonomy_decontam_r.tsv",
            sep = "\t",
            row.names = TRUE,
            quote = FALSE
            ) 
# still needs some modifications to be again compatible with qiime2 (check the R-imported file)

# Reference sequences table as fasta
writeXStringSet(refseq(ps_sub),
                "/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB1-2/rstudio/dna-sequences_r.fasta", 
                append = FALSE,
                compress = FALSE,
                compression_level = NA,
                format = "fasta"
                )

# Metadata file as tsv
write.table(sample_data(ps_sub),
            "metadata_r.tsv",
            sep = "\t",
            row.names = T,
            col.names = TRUE,
            quote = FALSE
            )

##############################################################################################

setwd("/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB1-2/microbiome-analyst/import-files")
getwd()
ps_ma <- subset_samples(ps_gen,
                        inflorescence != "" & inflorescence != "subdioecious"
                        ); ps_ma
asv_exp <- as(otu_table(ps_ma),
              "matrix"
              )
write.table(asv_exp,
            "gen_ma_inflorescence.txt",
            sep = "\t",
            quote = FALSE
            ) # ASV table as .tsv
```

```{r barplot, echo=FALSE}

# Microbial structure

# ps_plot <- subset_samples(ps_cannabis); ps_plot # select samples
# sample_names(ps_plot)
# nsamples(ps_plot)
# ps_mer <- merge_samples2(ps_plot, "cultivar")
ps_rel <- transform(ps_cannabis,
                    "compositional"
                    )
ps_agr <- aggregate_rare(ps_rel,
                         level = "Genus",
                         detection = .1/100,
                         prevalence = 70/100,
                         include.lowest = TRUE
                         )
ps_kno <- subset_taxa(ps_agr,
                      Genus != "Other"
                      )

mycolors <- colorRampPalette(brewer.pal(12, "Paired"))(8) # define colors
par(oma = c(12, 4, 4, 6) + 0.1) 
plot_bar(ps_kno,
         fill = "Genus",
         title = ""
         ) + 
  geom_bar(stat = "identity",
           position = "stack"
           ) + 
  scale_fill_manual(values = mycolors,
                    na.value = "grey50"
                    ) + 
  theme_bw() + 
  theme(axis.text.x = element_text(angle = 90)
        ) # + facet_wrap(~ cultivar, scales = "free_x")

setwd("/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB-all/rstudio")
ggsave("Core70only-genus-cultivar.png", 
       width = 40, 
       height = 20, 
       units = "cm", 
       dpi = 600
       )
# RColorBrewer::display.brewer.all() 
```

```{r rarefaction, echo = FALSE}

ps_cannabis # 5876 taxa and 458 samples

# normalize for alpha diversity
ps_raref <- rarefy_even_depth(ps_cannabis, 
                              sample.size = 8348, 
                              rngseed = 711, 
                              replace = FALSE, 
                              trimOTUs = TRUE
                              ) # depth = 6606 reads
ps_raref # 3343 taxa and 386 samples
sample_names(ps_raref)
ps_mer <- merge_samples(ps_raref, 
                        "cultivar"
                        ); ps_mer # n = 46, OK
summarize_phyloseq(ps_raref) # Average number of reads = 8348, Total number of reads = 3222328, Singletons = 33.3%
# ggrare(ps_raref, step = 10, color = "cultivar", se = FALSE) + facet_wrap(~ cultivar, scales = "free_x")

# normalize for core analysis
ps_cannabis_min <- filter_taxa(ps_cannabis, 
                               function(x) mean(x) < .1, 
                               TRUE
                               ); ps_cannabis_min # 5073 taxa and 458 samples
rmtaxa <- taxa_names(ps_cannabis_min)
alltaxa <- taxa_names(ps_cannabis)
mytaxa <- alltaxa[!alltaxa %in% rmtaxa]
ps_cannabis_1 <- prune_taxa(mytaxa,
                            ps_cannabis
                            ); ps_cannabis_1 # 803 taxa and 458 samples
```

```{r richness, echo=FALSE}

# Merged samples
# ps_set <- subset_samples(ps_raref, species2 != "" & species2 != "indica/sativa"); ps_set
richness <- estimate_richness(ps_raref, 
                              measures = c("Shannon", "Observed")
                              ); head(richness) # returns a data.frame
hist(richness$Shannon,
     main = "Shannon index",
     xlab = ""
     ) # check if the Shannon values are normally distributed

data <- cbind(sample_data(ps_raref),
              richness
              ); data
data.aov <- aov(Shannon ~ cultivar,
                data
                ); data.aov  # parametric
data.kt <- kruskal.test(Shannon ~ cultivar,
                        data
                        ); data.kt # non-parametric
data.wt <- pairwise.wilcox.test(data$Shannon, 
                                as.factor(data$cultivar), 
                                p.adjust.method = "bonf"
                                ); data.wt # for each pair of groups

# Significance
sig <- compare_means(Shannon ~ cultivar, 
                     data = data, 
                     p.adjust.method = "bonferroni"
                     )
sig_adj <- select(filter(sig,
                         p.adj < 0.05),
                  c(group1, 
                    group2, 
                    p.adj, 
                    p.signif)
                  )
class(sig_adj)
sample_pair <- data.frame(sig_adj$group1, 
                          sig_adj$group2
                          )

my_comparisons <- list()
for (idx in 1:nrow(sample_pair)) {
  g1 <- sample_pair[idx, "sig_adj.group1"]
  g2 <- sample_pair[idx, "sig_adj.group2"]
  my_comparisons[[idx]] <- c(g1, g2)
}
# print(my_comparisons)
str(my_comparisons)

symnum.args <- list(cutpoints = c(0, 0.0001, 0.001, 0.01, 0.05, 1),
                    symbols = c("****", "***", "**", "*", "ns")
                    ) # add significance comparisons
# symnum.args <- list(cutpoints = c(1), symbols = c("*")) # use a threshold

# ps_sample <- subset_samples(ps_raref, species.1 != "" & inflorescence == "dioecious") # select samples

(plot_richness(ps_raref, 
               x = "cultivar", 
               color = "cultivar", 
               measures = c("Shannon")
               ) +
  xlab("") +
  geom_boxplot(aes(fill = cultivar), alpha = 0.2) + 
  stat_compare_means(method = "wilcox.test",
      comparisons = as.list(my_comparisons),
      symnum.args = symnum.args,
      hide.ns = TRUE,
      na.rm = T
    ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) #, vjust = 0.5, hjust = 0))
)
setwd("/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB-all/rstudio")
ggsave("shannon-sig.png", 
       width = 40, 
       height = 20, 
       units = "cm"
       )

# Replicates
cultivars <- unique(sample_data(ps_raref)$cultivar); cultivars # 46 cultivars
for (c in cultivars) {
  # ps_s <- subset_samples(ps_raref, cultivar == c)
  physeq <- ps_raref
  oldDF <- as(sample_data(physeq), "data.frame")
  newDF <- subset(oldDF, cultivar == c)
  if (class(physeq) == "sample_data") {
      ps_s <- sample_data(newDF)
  }
  else {
      sample_data(physeq) <- sample_data(newDF)
      ps_s <- physeq
  }
  attributes(ps_s)

  richness <- estimate_richness(ps_s, 
                                measures = "Shannon"
                                ) # returns a data.frame
  p <- plot_richness(ps_s, 
                     x = "replicates", 
                     color = "cultivar", 
                     measures = "Shannon") + 
    labs(x = "", 
         y = "Alpha Diversity Measure\n") +
    theme_bw()
  print(p)
}
```

```{r dominant, core and rare species, echo = FALSE}

# file:///C:/Users/pongo/Downloads/Krona2.html?dataset=0&node=2&collapse=false&color=false&depth=4&font=11&key=true # Kronaplot

dom_sum <- dominant_taxa(ps_cannabis, 
                         level = "Phylum", 
                         group = NULL
                         ); dom_sum # dominant taxonomic group overview and percent 
sample_dom <- dominant(ps_cannabis, 
                       level = "Genus"
                       ); sample_dom # retrieves the dominant taxonomic group of each sample
dom_ab <- dominance(ps_cannabis, 
                    index = "relative", 
                    relative = TRUE, 
                    aggregate = TRUE
                    ); dom_ab # gives the relative abundance of the most abundant SPECIES per sample
summary(dom_ab) # Mean != with tax_glom

ps_samples <- subset_samples(ps_cannabis, 
                             cultivar == "Fedora 17"
                             ); ps_samples

ps_core <- core(ps_gen, 
                detection = .1/100, 
                prevalence = 70/100, 
                include.lowest = FALSE
                ); ps_core # phyloseq object with only prevalent SPECIES
core_tax <- core_members(ps_gen, 
                         detection = .1/100, 
                         prevalence = 70/100, 
                         include.lowest = FALSE
                         ); core_tax # lists prevalent SPECIES
core_ab <- core_abundance(ps_cannabis, 
                          detection = .1/100, 
                          prevalence = 75/100
                          );core_ab # relative proportion of the core SPECIES
summary(core_ab) # Mean != with tax_glom
sd(core_ab)

ps_cultivar <- merge_samples(ps_samples, 
                             "inflorescence"
                             ); ps_cultivar

ps_rare <- rare(ps_cannabis, 
                detection = .1/100, 
                prevalence = 75/100, 
                include.lowest = FALSE
                ); ps_rare # phyloseq object with only rare SPECIES.
rare_tax <- rare_members(ps_gen, 
                         detection = .1/100, 
                         prevalence = 50/100, 
                         include.lowest = FALSE
                         ); rare_tax # lists rare SPECIES
rare_ab <- rare_abundance(ps_gen, 
                          detection = .1/100, 
                          prevalence = 1/31, 
                          include.lowest = FALSE
                          ); rare_ab # relative proportion of rare SPECIES
summary(rare_ab) # Mean != with tax_glom

ps_rel <- transform(ps_gen, 
                    "compositional"
                    )
ps_taxa <- subset_taxa(ps_gen, 
                       Genus == "Paracoccus"
                       ); ps_taxa
a <- abundances(ps_taxa)
sum(a)
nsamples(ps_taxa)
```

```{r beta-diversity, echo=FALSE}

ps_new <- subset_samples(ps_cannabis); ps_new # select samples
ps_mer <- merge_samples(ps_new,
                        "cultivar"
                        )

ps_css <- phyloseq_transform_css(ps_mer,
                                 norm = TRUE,
                                 log = TRUE
                                 ) # correct uneven sequencing depth using cumulative sum scaling (CSS) instead of rarefaction

ps_MDS <- ordinate(ps_css,
                   "MDS", 
                   "bray",
                   trymax = 1000
                   ); ps_MDS # Non-metric MultiDimenstional Scaling
plot_ordination(ps_css,
                ps_MDS, 
                type = "samples",
                color = "cultivar",
                shape = "",
                label = "cultivar"
                ) + 
  geom_point(size = 1) + 
  theme_bw() +
  ggtitle("MDS + Bray Curtis") +
  stat_ellipse(type = "norm",
               linetype = 1,
               level = 0.95
               )
ggsave("mds-bray-cultivar.png",
       width = 40,
       height = 20,
       units = "cm"
       )

dist_matrix_t <- t(data.frame(otu_table(ps_css))) # transpose the ASV table
df_css <- data.frame(sample_data(ps_css))
ps_css_bray <- vegdist(dist_matrix_t,
                       method = "bray"
                       ) # generate a bray distance object
adonis2(ps_css_bray ~ species2,
        data = df_css
        )

css_table <- as(otu_table(ps_css),
                "matrix"
                )
if(taxa_are_rows(ps_css)){css_table <- t(css_table)} # transpose if necessary
css_table_df = as.data.frame(css_table) # coerce to data.frame
write.csv(css_table_df,
          file = "css_table.csv"
          )
css_table_MPC <- read.table("css_table.txt",
                            header = TRUE
                            )

pairwise.adonis(css_table_MPC[,2:xx],
                as.factor(css_table_MPC$"cultivar"),
                sim.function = "vegdist",
                sim.method = "bray",
                p.adjust.m = "fdr"
                ) # pairwise comparisons
```

```{r subsetting taxonomic levels, echo = FALSE}

# df <- psmelt(ps_cannabis)
# kable(head(df))

# Working with different taxonomic levels
ps_spe <- phyloseq::tax_glom(ps_cannabis,
                             "Species",
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_spe # 163 taxa
# taxa_names(ps_spe) <- tax_table(ps_spe)[,"Species"] # Assign taxa names
unique(tax_table(ps_spe)[,"Species"]) # 156 taxa

ps_gen <- phyloseq::tax_glom(ps_cannabis,
                             "Genus",
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_gen # 678 taxa
# taxa_names(ps_gen) <- tax_table(ps_gen)[,"Genus"] # Assign taxa names
unique(tax_table(ps_gen)[,"Genus"]) # 678 taxa

ps_fam <- phyloseq::tax_glom(ps_cannabis,
                             "Family",
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_fam # 364 taxa
# taxa_names(ps_fam) <- tax_table(ps_fam)[,"Family"] # Assign taxa names
unique(tax_table(ps_fam)[,"Family"]) # 364 taxa

ps_ord <- phyloseq::tax_glom(ps_cannabis,
                             "Order", 
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_ord # 223
# taxa_names(ps_ord) <- tax_table(ps_ord)[,"Order"] # Assign taxa names
unique(tax_table(ps_fam)[,"Order"]) # 220 taxa 

ps_cla <- phyloseq::tax_glom(ps_cannabis,
                             "Class", 
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_cla # 96 taxa
# taxa_names(ps_cla) <- tax_table(ps_cla)[,"Class"] # Assign taxa names
unique(tax_table(ps_cla)[,"Class"]) # 96 taxa

ps_phy <- phyloseq::tax_glom(ps_cannabis,
                             "Phylum", 
                             NArm = TRUE
                             ) # NArm removes anything that is unclassified
ps_phy # 37 taxa 
taxa_names(ps_phy) <- tax_table(ps_phy)[,"Phylum"] # Assign taxa names
unique(tax_table(ps_phy)[,"Phylum"]) # 37 taxa
```

```{r export cytoscape and source tracker, echo=FALSE}

# use the agglomerated phyloseq object to (transform into compositional data and filter reads by detection level (d = 0.1), (merge by cultivar), and compute the core microbiome (0.1; 0.75)

# ps_norm <- filter_taxa(ps_core, function(x) sum(x > 3) > (0.2 * length(x)), TRUE); ps_norm # avoid spurious correlations by removing taxa not seen more than 3 times in at least 20% of the samples
ps_sample <- subset_samples(ps_cannabis,
                            cultivar != ""
                            ); ps_sample
ps_cultivar <- merge_samples(ps_cannabis,
                             "cultivar"
                             ); ps_cultivar # 10 taxa and 31 samples
# ps_rel <-  transform(ps_norm, "compositional")

asv_cyt <- otu_table(ps_cannabis_1)
asv_cyt <- make_biom(data = asv_cyt)
write_biom(asv_cyt,
           "asv_cyt_gen.1_r.biom"
           )# ASV table as biom

taxonomy_cyt <- as(tax_table(ps_cannabis_1),
                   "matrix"
                   )
taxonomy_cyt <- as.data.frame(taxonomy_cyt)
write.table(taxonomy_cyt,
            "taxonomy_cyt_gen.1_r.tsv",
            sep = "\t",
            row.names = TRUE,
            quote = FALSE
            ) # Taxonomy table as tsv

##########################################################################################################################################
setwd("/home/jmachadofreitas/Documents/Carolina/22-amp-seq-seedlibrary/metabarcoding/results/anchored/CB1-2/qiime1-st")
getwd()

asv_exp <- as(otu_table(ps_cultivar),
              "matrix"
              )
asv_t <- t(asv_exp)
write.table(asv_t,
            "asv_cultivar_st.txt",
            sep = "\t",
            quote = FALSE
            ) # ASV table as .tsv
```

```{r heatmaps, echo = FALSE}

ps_c75 <- subset_samples(ps_raref,
                         harvest.days != ""
                         )
ps_c75 <- phyloseq::merge_samples(ps_c75,
                                  "harvest.days"
                                  )
ps_c75 <- microbiome::transform(ps_c75,
                                "compositional"
                                )
ps_c75 <- aggregate_rare(ps_c75,
                         level = "Genus",
                         NArm = TRUE,
                         detection = .1/100,
                         prevalence = 75/100
                         )

days <- factor(sample_data(ps_c75)$harvest.days,
               levels = sort(unique(sample_data(ps_c75)$harvest.days))
               )

heat_core <- (
  ps_c75 %>%
  microbiome::plot_composition(
    sample.sort = "neatmap",
    otu.sort = "neatmap",
    plot.type = "heatmap") +
    ylab("Germination rate") +
    theme(axis.text.y = element_text(size = 8,
                                     hjust = 1),
          axis.ticks.y = element_blank(),
          axis.text.x = element_text(size = 8,
                                     hjust = 1,
                                     angle = 30),
          legend.text = element_text(size = 8)) +
    scale_fill_gradient(limits=c(0, 1),
                        low = "white",
                        high = "#9c0404",
                        name = ""
                        )
)
print(heat_core)
```

```{r source tracker echo=FALSE}

metadata <- read.table("source-tracker/metadata_st_Fibrimon_Fedora17_rep.txt",
                       sep = '\t'
                       )

otus <- read.csv("source-tracker/asv_st_rep_r.txt")
otus <- t(as.matrix(otus))

common.sample.ids <- intersect(rownames(metadata),
                               rownames(otus)
                               )
otus <- otus[common.sample.ids,] # extract samples in common between tables
metadata <- metadata[common.sample.ids,] # extract samples in common between tables

# Check if overlapping samples exist
if(length(common.sample.ids) <= 1) {
  message <- paste(sprintf("Error: there are %d sample ids in common between the metadata file and data table"))
  stop(message)
}

# Extract source environments and source/sink indexes
train.ix <- which(metadata$SourceSink == "source")
test.ix <- which(metadata$SourceSink == "sink")
envs <- metadata$Env
if(is.element("Description",
              colnames(metadata))) desc <- metadata$Description

source("source-tracker/SourceTracker.R") # load SourceTracker package

alpha1 <- alpha2 <- 0.05 # skip tuning, run this instead
st <- sourcetracker(otus[train.ix,],
                    envs[train.ix]
                    ) # train SourceTracker object on training data

# Estimate source proportions in test data
results <- predict(st,
                   otus[test.ix,],
                   alpha1 = alpha1,
                   alpha2 = alpha2
                   )
results$proportions_sd
results$samplenames
results$proportions
results$draws

# Estimate leave-one-out source proportions in training data 
results.train <- predict(st,
                         alpha1 = alpha1,
                         alpha2 = alpha2
                         )

# plot results
labels <- sprintf('%s %s',
                  envs,
                  desc
                  )
plot(results,
     labels[test.ix],
     type ='pie'
     
     )

# other plotting functions
plot(results,
     labels[test.ix],
     type = 'bar'
     )

##############################################################################################
install.packages("ggmosaic")
install.packages("vcd")
library(ggmosaic)

# Create variables
Sample <- c("Fibrimon", "Unknown")
Beniko <- c(0.12, 0.98)
df <- data.frame(Sample, Beniko); df # Join the variables to create a data frame
thm <- theme_minimal() +
  theme(text = element_text(size = 16)) +
  ggplot() +
  geom_mosaic(aes(x = product(Sample),
                    fill = Beniko)) +
  scale_fill_brewer(palette = "Blues",
                    direction = -1)

thm + labs(x = NULL, y = "Beniko")
```
