---
title: "Preprocessing"
author: "Carolina Lobato"
date: "27/02/2023"
output: html_document
---

***Create, preprocess, and save PhyloSeq objects and tables***

# Load Libraries

```{r include = FALSE}
library(Biostrings) 
library(phyloseq) 
library(decontam) 
library(microbiome) 
library(microbiomeutilities) 
library(ranacapa) 
library(tidyverse) 
```

# Import Data ======================================================

```{r, echo = FALSE}
## Import metadata file
metadata <- read.table(
  "metadata/metadata.csv",
  sep = ";",
  header = T,
  fill = T,
  row.names = "OTUID",
  na.strings = "")
metadata$replicates <- row.names(metadata)
metadata <- sample_data(metadata)

## Import feature table
asv_tab <- read.table(
  "outputs/qiime2/feature-table.txt",
                        header = T,
                        row.names = "OTUID")
asv_tab <- otu_table(asv_tab, taxa_are_rows = T)

## Import taxonomy file
tax_tab <- read.table(
  "outputs/qiime2/16S-taxonomy-vsearch.tsv",
  sep = "\t",
  header = T,
  fill = T,
  row.names = "OTUID")
tax_tab <- as.matrix(tax_tab)
tax_tab <- tax_table(tax_tab)

## Import reference sequences
rs_table <- Biostrings::readDNAStringSet(
  "outputs/qiime2/dna-sequences.fasta",
  format = "fasta",
  nrec = -1L,
  skip = 0L,
  seek.first.rec = F,
  use.names = T,
  with.qualities = F)

## Import rooted tree
rtree <- read_tree("outputs/qiime2/rooted-tree.nwk")

## Convert to phyloseq object
ps <- merge_phyloseq(
  metadata,
  asv_tab,
  tax_tab,
  rs_table,
  rtree)
summarize_phyloseq(ps)
ps0 <- ps
save(ps0,
     file = "outputs/r/setup/ps0.RData")
```

# Decontam =========================================================

```{r, echo = FALSE}
load("outputs/r/setup/ps0.RData")

## Inspect
df <- as.data.frame(sample_data(ps0))
df$Library_Size <- sample_sums(ps0)
df <- df[order(df$Library_Size),]
df$Index <- seq(nrow(df))

ggplot(data = df, 
       aes(x = Index,
           y = Library_Size,
           color = type)) +
  geom_point()

## By Prevalence
sample_data(ps0)$is.neg <- sample_data(ps0)$type == "NC"

contamdf.prev <- isContaminant(ps0, 
                               method = "prevalence",
                               batch = "batch",
                               batch.combine = "fisher", 
                               neg = "is.neg")
table(contamdf.prev$contaminant)

## New phy-obj
ps0_nc <- prune_taxa(!contamdf.prev$contaminant, ps0)
save(ps0_nc,
     file = "outputs/r/setup/ps0-decontam.RData")

# Check for the number of reads and ASVs
contam <- prune_taxa(contamdf.prev$contaminant, ps0)
View(data.frame(tax_table(contam)))

samples <- prune_samples(sample_data(contam)$type == "sample", contam)

sum(sample_sums(samples))/sum(sample_sums(ps0)) * 100 # 6.5%

## PA phy-obj for NTC and true samples
ps.pa <- transform_sample_counts(ps0, function(abund) 1 * (abund > 0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$type == "NC", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$type != "NC", ps.pa)

## Prevalence data.frame for positive and negative samples
df.pa <- data.frame(pa.pos = taxa_sums(ps.pa.pos),
                    pa.neg = taxa_sums(ps.pa.neg),
                    contaminant = contamdf.prev$contaminant)
## Plot
ggplot(data = df.pa, 
       aes(x = pa.neg, 
           y = pa.pos, 
           color = contaminant)) +
  geom_point() +
  xlab("Prevalence (NTC)") +
  ylab("Prevalence (True Samples)")
```

# Clean ===========================================================

```{r, echo = FALSE}
load("outputs/r/setup/ps0-decontam.RData") # ps0_nc

ps_clean <- subset_samples(ps0_nc, type != "M")
ps_clean <- subset_samples(ps_clean, type != "NC")
ps_clean <- subset_samples(ps_clean, replicates != "C34_01B")
ps_clean <- prune_samples(sample_sums(ps_clean) > 0, ps_clean)
ps_clean <- subset_taxa(ps_clean, Domain != "d__Archaea")
ps_clean <- prune_taxa(taxa_sums(ps_clean) > 0, ps_clean)
ps_clean # 5297 taxa and 451 samples

sort(sample_sums(ps_clean))
summary(sample_sums(ps_clean)) # mean reads/sample = 82033
summarize_phyloseq(ps_clean) # Average reads = 82033, Total reads = 36996902, Singletons = 0
taxa_names(ps_clean)[1:5]
sample_data(ps_clean)$total_reads <- sample_sums(ps_clean)

## Replace unknown names
tax_table(ps_clean)[tax_table(ps_clean) == "d__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "p__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "c__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "o__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "f__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "f__Unknown_Family"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "g__uncultured"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "g__possible_genus_04"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "s__uncultured_bacterium"] <-
  ""
tax_table(ps_clean)[tax_table(ps_clean) == "s__metagenome"] <-
  ""
tax_tab <- phyloseq::tax_table(ps_clean); tax_tab

## Number ASVs
taxa_names(ps_clean) <- paste(
  "ASV ", 1:length(taxa_names(ps_clean)),
  ": ",
  taxa_names(ps_clean),
  sep = "")
taxa_names(ps_clean)[1:5]

## Remove tax level indicator in the beginning of each taxa
tax_table(ps_clean)[, colnames(tax_table(ps_clean))] <- 
  gsub(tax_table(ps_clean)[, colnames(tax_table(ps_clean))],
  pattern = "[a-z]__",
  replacement = "")

ps <- ps_clean
save(ps, file = "outputs/r/setup/ps.RData")
conversion <- taxa_names(ps_clean); conversion
write.table(conversion, "outputs/r/setup/tax-conv.csv")

## Replace ASV code with best hit
bh <- microbiomeutilities::format_to_besthit(ps_clean)
taxa_names(bh) <- sub(": [a-z0-9]*:[a-z_]{0,3}", ": ", taxa_names(bh))
taxa_names(bh)[1:5]
save(bh, file = "outputs/r/setup/bh.RData")

## Get number of taxa across different levels:
rank_names(bh)
D <- as.data.frame(table(tax_table(bh)[,"Domain"]))
P <- as.data.frame(table(tax_table(bh)[,"Phylum"]))
C <- as.data.frame(table(tax_table(bh)[,"Class"]))
O <- as.data.frame(table(tax_table(bh)[,"Order"]))
Fa <- as.data.frame(table(tax_table(bh)[,"Family"]))
G <- as.data.frame(table(tax_table(bh)[,"Genus"]))
S <- as.data.frame(table(tax_table(bh)[,"Species"]))

## Save final tables
taxtable <- bh %>%
  tax_table %>%
  as.matrix %>%
  data.frame

write.table(taxtable,
            "outputs/r/setup/bh-tax-table.csv",
            sep = ",",
            row.names = F,
            quote = F)

otutable <- bh %>%
  otu_table %>%
  as.matrix %>%
  data.frame

write.table(otutable,
            "outputs/r/setup/bh-otu-table.tsv",
            sep = "\t",
            row.names = T,
            quote = F)
```

# Rarefy Samples ===================================================

```{r, echo = FALSE}
load("outputs/r/setup/bh.RData")

bh_raref <- phyloseq::rarefy_even_depth(
  bh,
  sample.size = 5607,
  rngseed = 711,
  replace = F,
  trimOTUs = T)
bh_raref # 2795 taxa and 402 samples
sample_names(bh_raref)
bh_mer <- phyloseq::merge_samples(bh_raref, "genotype"); bh_mer # n = 46, OK
summarize_phyloseq(bh_raref) # Average reads = 5607, Total reads = 2254014, Singletons = 985

save(bh_raref, file = "outputs/r/setup/bh-raref.RData")

ranacapa::ggrare(bh_raref,
       step = 10,
       color = "chemotype2",
       se = F) +
  theme(legend.position = "none") +
  facet_wrap(~chemotype2, scales = "free_x", ncol = 3)
```
