---
title: "Rare taxa"
author: "Carolina Lobato"
date: "2023-10-10"
output: html_document
---

# Load libraries

```{r, echo = FALSE}
library(phyloseq)
library(microbiome)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(colorspace)
library(RColorBrewer)
library(SpiecEasi)
library(igraph)
library(ggnet)
library(network)
library(ggnetwork)
library(GGally)
library(intergraph)
library(ggpubr)
library(RMThreshold)
```

# List of core and rare taxa

```{r, echo = FALSE}
load("outputs/r/204_setup/bh.RData")
meta <- meta(bh)

# Remove taxa not seen more than 5 times in at least 0.1% of the samples: 
bh_pruned <- filter_taxa(bh, function(x) sum(x > 5) > (0.001*length(x)), TRUE); bh_pruned # 2612

# save(bh_pruned, file = "outputs/r/308_rare_taxa/bh-5-01.RData")

# bh_china <- subset_samples(bh_pruned, cultivar == "NA BG China")
# bh_china <- prune_taxa(taxa_sums(bh_china) > 0, bh_china); bh_china
# taxa_names(bh_china)

bh_core <- microbiome::core(bh_pruned,
                            detection = 0.1,
                            prevalence = 0.75, # min 339 samples
                            include.lowest = TRUE) # phyloseq object

gen.core <- taxa_names(bh_core); gen.core # 9 ASVs
# ASV 2062: Pelomonas
# ASV 2280: Ralstonia
# ASV 2328: Ralstonia
# ASV 2389: Burkholderia_sp.
# ASV 3007: Bacillus  
# ASV 3198: Pseudomonas_sp.
# ASV 3353: Pantoea_agglomerans
# ASV 3464: Enhydrobacter
# ASV 4544: Rhodococcus_erythropolis

bh_rare <- microbiome::rare(bh,
                            detection = 0.1,
                            prevalence = 0.2, # max 90 samples
                            include.lowest = TRUE) # phyloseq object

gen.rare <- taxa_names(bh_rare); gen.rare # 2560 ASVs


sample_sums(bh) # sum of all reads observed in each sample
sample_sums(bh) # sum of all core reads observed in each sample
sample_sums(bh_rare) # sum of all rare reads observed in each sample
taxa_sums(bh) # sum of all reads observed in each taxa
taxa_sums(bh_core) # sum of all reads observed in each core taxa
taxa_sums(bh_rare) # sum of all reads observed in each rare taxa
summarize_phyloseq(bh) # total number of reads = 36996923
summarize_phyloseq(bh_core) # total number of core reads = 20820554
summarize_phyloseq(bh_rare) # total number of rare reads = 95474

# save(gen.core, file = "outputs/r/308_rare_taxa/gen-core.RData")
# save(gen.rare, file = "outputs/r/308_rare_taxa/gen-rare.RData")
```

# Prevalence and %RA of core and rare taxa

```{r, echo = FALSE}
load("outputs/r/204_setup/bh.RData")
load("outputs/r/308_rare_taxa/bh-5-01.RData")

asv <- rownames_to_column(as.data.frame(otu_table(bh)))
taxo <- tax_table(bh) 
meta <- microbiome::meta(bh) 

# presence-absence data
ASV_PA <- 1 * ((asv > 0) == 1)

# occupancy calculation
ASV_prev <- rowSums(ASV_PA) / ncol(ASV_PA)

# relative abundance
ASV_rel <- apply(vegan::decostand(asv, 
                                  method = "total", 
                                  MARGIN = 2), 1, mean)

# combine occupancy and RA of each ASV in a table
ASV_prev_rel <- rownames_to_column(as.data.frame(cbind(ASV_prev, ASV_rel)), "ASV")

ASV_prev_rel_taxo <- rownames_to_column(as.data.frame(taxo),
                                        "ASV") 
ASV_prev_rel_taxo <- merge(ASV_prev_rel_taxo, 
                           ASV_prev_rel, 
                           by = "ASV")

# merge with metadata
ASV1 <- rownames_to_column(as.data.frame(t(asv)),
                           "replicates")

ASV_use1 <- merge(meta, 
                  ASV1, 
                  by = "replicates")

# keep only the one feature column for collapsing table
ASV_use1_1 <- ASV_use1[-c(1:3, 5:17)]

ASV_cultivar <- ASV_use1_1 %>% 
  group_by(cultivar) %>% 
  summarise_each(list(sum))

# make a ASV P/A table
ASV_cultivar_PA <- 1 * ((ASV_cultivar > 0) == 1)

# merge prevalence, RA and taxa with the nº of observations
ASV_obs_cultivar <- data.frame(colSums(ASV_cultivar_PA)) %>%
  na.omit() %>% 
  rownames_to_column("ASV")
names(ASV_obs_cultivar)[names(ASV_obs_cultivar) == "colSums.ASV_cultivar_PA."] <- 'Obs'

ASV_prev_rel_taxo_obs_cultivar <- merge(ASV_prev_rel_taxo, 
                                        ASV_obs_cultivar, 
                                        by = "ASV")

# save(ASV_prev_rel_taxo_obs_cultivar,
#      file = "outputs/r/308_rare_taxa/prev-rel-taxo-obs-cultivar.RData")
```

## Scatter plot prep

```{r}
load("outputs/r/308_rare_taxa/prev-rel-taxo-obs-cultivar.RData") # ASV_prev_rel_taxo_obs_cultivar
load("outputs/r/308_rare_taxa/gen-core.RData") # gen.core
load("outputs/r/308_rare_taxa/gen-rare.RData") # gen.rare

ASV_prev_rel_taxo_obs_cultivar_cr <-
  ASV_prev_rel_taxo_obs_cultivar %>% 
  mutate(core_asv = ASV %in% gen.core) %>% 
  mutate(rare_asv = ASV %in% gen.rare) %>% 
  mutate(classification = case_when(
    (.$core_asv == TRUE) ~ "Core",
    (.$rare_asv == TRUE) ~ "Rare")) %>% 
  replace_na(list(classification = "Others")) %>%
  mutate(ASV_rel_p = ASV_rel * 100, 
         ASV_prev_p = ASV_prev * 100)

ASV_prev_rel_taxo_obs_cultivar_cr$classification <-
  ordered(ASV_prev_rel_taxo_obs_cultivar_cr$classification, 
          levels = c("Core", "Rare", "Others"))

# save(ASV_prev_rel_taxo_obs_cultivar_cr,
#      file = "outputs/r/308_rare_taxa/prev-rel-taxo-obs-cultivar-cr.RData")
```

### Scatter Plot

```{r}
load("outputs/r/308_rare_taxa/prev-rel-taxo-obs-cultivar-cr.RData") # ASV_prev_rel_taxo_obs_cultivar_cr

color_cr = c(Core = "#C7522B", Rare = "#C7D6B2", Others = "#D3D3D3")

p_cr <- 
  ggplot(
    ASV_prev_rel_taxo_obs_cultivar_cr,
    aes(x = ASV_rel_p, 
        y = ASV_prev_p, 
        color = classification)) +
      geom_point(size = 3, 
                 alpha = 0.7) + 
  # geom_errorbar(aes(xmin = ASV_prev_p - sd(ASV_prev_p),
  #                   xmax = ASV_prev_p + sd(ASV_prev_p)),
  #                   width = 0.2,
  #                   position = "identity") +
  # ggrepel::geom_text_repel(data = subset(ASV_prev_rel_taxo_obs_cultivar_cr, 
  #                         ASV_rel_p > 0.1 & ASV_prev_p > 75),
  #           aes(ASV_rel_p, 
  #               ASV_prev_p, 
  #               label = Genus), 
  #           size = 3,
  #           show.legend = F) +
  geom_hline(yintercept = 75,
             linetype = "dashed", 
             color = "black") +
  geom_hline(yintercept = 1,
             linetype = "dashed", 
             color = "black") +
  geom_vline(xintercept = 0.1,
             linetype = "dotted",
             color = "black") +
      theme_classic() +
      theme(
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = "right",
        legend.direction = "vertical",
        panel.spacing = unit(0.1, "lines")) +
      scale_x_log10(breaks = c(10, 0.1, 0.001, 0.00001),
                    labels = c(10, 0.1, 0.001, 0.00001)) +
      scale_color_manual(values = color_cr)	+
      labs(x = "Relative Abundance (%)", 
           y = "ASV Prevalence (%)", 
           color = "")
p_cr

# ggsave("outputs/visual/r/308_rare_taxa/gen-core-rare-1.png",
#         width = 20,
#         height = 18,
#         units = "cm",
#         scale = 1,
#         dpi = 600)
```

## Subset

```{r}
load("outputs/r/204_setup/bh.RData")
load("outputs/r/308_rare_taxa/bh-5-01.RData")

Bacteria_seed_china <- subset_samples(bh, cultivar == "NA BG China")
Bacteria_seed_china <- prune_taxa(taxa_sums(Bacteria_seed_china) > 0, Bacteria_seed_china); Bacteria_seed_china # 532 taxa / 392 taxa
china.readsums <- data.frame(nreads = sort(taxa_sums(Bacteria_seed_china),
                                           TRUE), 
                             sorted = 1:ntaxa(Bacteria_seed_china), 
                             type = "OTU")
## Remove OTUs with low counts
summary(taxa_sums(Bacteria_seed_china))
Bacteria_seed_china.p <-  prune_taxa(taxa_sums(Bacteria_seed_china) > 4, 
                                     Bacteria_seed_china) # 379 taxa and 10 samples

Bacteria_seed_hgr <- subset_samples(bh, cultivar == "Hohenthurmer G. R.")
Bacteria_seed_hgr <- prune_taxa(taxa_sums(Bacteria_seed_hgr) > 0, Bacteria_seed_hgr); Bacteria_seed_hgr # 344 taxa and 10 samples

hgr.readsums <- data.frame(nreads = sort(taxa_sums(Bacteria_seed_hgr), TRUE), 
                             sorted = 1:ntaxa(Bacteria_seed_hgr), 
                             type = "OTU")
## Remove OTUs with low counts
summary(taxa_sums(Bacteria_seed_hgr))
Bacteria_seed_hgr.p <-  prune_taxa(taxa_sums(Bacteria_seed_hgr) > 3, 
                                     Bacteria_seed_hgr) # 227 taxa and 10 samples

Bacteria_seed_masterk <- subset_samples(bh, cultivar == "Master Kush")
Bacteria_seed_masterk <- prune_taxa(taxa_sums(Bacteria_seed_masterk) > 0, Bacteria_seed_masterk); Bacteria_seed_masterk # 453 taxa and 10 samples

masterk.readsums <- data.frame(nreads = sort(taxa_sums(Bacteria_seed_masterk), TRUE), 
                             sorted = 1:ntaxa(Bacteria_seed_masterk), 
                             type = "OTU")
## Remove OTUs with low counts
summary(taxa_sums(Bacteria_seed_masterk))
Bacteria_seed_masterk.p <-  prune_taxa(taxa_sums(Bacteria_seed_masterk) > 4, 
                                     Bacteria_seed_masterk) # 334 taxa and 10 samples

# save(Bacteria_seed_china.p, file = "outputs/r/308_rare_taxa/bac-china.RData")
# save(Bacteria_seed_hgr.p, file = "outputs/r/308_rare_taxa/bac-hgr.RData")
# save(Bacteria_seed_masterk.p, file = "outputs/r/308_rare_taxa/bac-masterk.RData")
```

## Create pulsar refit

```{r}
load("outputs/r/308_rare_taxa/bac-china.RData") # Bacteria_seed_china.p
load("outputs/r/308_rare_taxa/bac-hgr.RData") # Bacteria_seed_hgr.p
load("outputs/r/308_rare_taxa/bac-masterk.RData") # Bacteria_seed_masterk.p

se.mb.china_01_10_20 <- spiec.easi(
  Bacteria_seed_china,
  method = "mb",
  sel.criterion = "bstars",
  lambda.min.ratio = 0.1,
  nlambda = 10,
  pulsar.params = list(rep.num = 20, ncores = 4, seed = 54321),
  verbose = T)

getStability(se.mb.china_01_10_20) # 0.01850903
sum(getRefit(se.mb.china_01_10_20)) / 2 # 823
se.mb.china_01_10_20$select$stars$summary # 0.01850903

se.mb.hgr_01_10_20 <- spiec.easi(
  Bacteria_seed_hgr.p,
  method = "mb",
  sel.criterion = "bstars",
  lambda.min.ratio = 0.1,
  nlambda = 10,
  pulsar.params = list(rep.num = 20, ncores = 4, seed = 54321),
  verbose = T)

getStability(se.mb.hgr_01_10_20) # 0.02107598
sum(getRefit(se.mb.hgr_01_10_20)) / 2 # 285
se.mb.hgr_01_10_20$select$stars$summary # 0.02107598

se.mb.masterk_01_10_20 <- spiec.easi(
  Bacteria_seed_masterk.p,
  method = "mb",
  sel.criterion = "bstars",
  lambda.min.ratio = 0.1,
  nlambda = 10,
  pulsar.params = list(rep.num = 20, ncores = 4, seed = 54321),
  verbose = T)

getStability(se.mb.masterk_01_10_20) # 0.02052939
sum(getRefit(se.mb.masterk_01_10_20)) / 2 # 605
se.mb.masterk_01_10_20$select$stars$summary # 0.02052939

# save(se.mb.china_01_10_20, file = "outputs/r/308_rare_taxa/se-china.RData")
# save(se.mb.hgr_01_10_20, file = "outputs/r/308_rare_taxa/se-hgr.RData")
# save(se.mb.masterk_01_10_20, file = "outputs/r/308_rare_taxa/se-masterk.RData")
```

### Calculate correlation thresholds

```{r}
load("outputs/r/308_rare_taxa/se-china.RData") # se.mb.china_01_10_20
betaMat_china <- as.matrix(symBeta(getOptBeta(se.mb.china_01_10_20)))
summary(betaMat_china)
 
positive_china <- length(betaMat_china[betaMat_china > 0.35])/2; positive_china # 174
negative_china <- length(betaMat_china[betaMat_china < -0.35])/2; negative_china # 2
total_china <- length(betaMat_china[betaMat_china != 0.35])/2; total_china # 71820.5
# save(betaMat_china, file = "outputs/r/308_rare_taxa/betmat-china.RData")

load("outputs/r/308_rare_taxa/se-hgr.RData") # se.mb.hgr_01_10_20
betaMat_hgr <- as.matrix(symBeta(getOptBeta(se.mb.hgr_01_10_20)))
summary(betaMat_hgr)

positive_hgr <- length(betaMat_hgr[betaMat_hgr > 0.3])/2; positive_hgr # 42
negative_hgr <- length(betaMat_hgr[betaMat_hgr <- 0.3])/2; negative_hgr # 0
total_hgr <- length(betaMat_hgr[betaMat_hgr != 0.3])/2; total_hgr # 25764.5
# save(betaMat_hgr, file = "outputs/r/308_rare_taxa/betmat-hgr.RData")

load("outputs/r/308_rare_taxa/se-masterk.RData") # se.mb.masterk_01_10_20
betaMat_masterk <- as.matrix(symBeta(getOptBeta(se.mb.masterk_01_10_20)))
summary(betaMat_masterk)

positive_masterk <- length(betaMat_masterk[betaMat_masterk > 0.3])/2; positive_masterk # 81
negative_masterk <- length(betaMat_masterk[betaMat_masterk <- 0.3])/2; negative_masterk # 0
total_masterk <- length(betaMat_masterk[betaMat_masterk != 0.3])/2; total_masterk # 55778
# save(betaMat_masterk, file = "outputs/r/308_rare_taxa/betmat-masterk.RData")
```

### RMThreshold: China

```{r}
load("outputs/r/308_rare_taxa/betmat-china.RData")
dim(betaMat_china)
connections_china <- rm.connections(betaMat_china) 
val_china <- rm.matrix.validation(betaMat_china) 
str(val_china)
# Eigenvalue distribution = Wigner semi-circle OK
# Spacing = Wigner-Dyson distribution OK but use discard.zeros = TRUE


thr_china <- RMThreshold::rm.get.threshold(betaMat_china, discard.zeros = T)
str(thr_china)

denoise_china <- rm.denoise.mat(betaMat_china, 
                                threshold = 0.35) # Number of non-zero matrix elements reduced from 1646 to 352 
clean_china = rm.discard.zeros(denoise_china)	# 113 zero-rows and 113 zero-columns removed.
connections_clean_china <- rm.connections(clean_china)

g_china  <- graph.adjacency(clean_china, mode = "undirected")
clusters(g_china)

positive_china <- length(clean_china[clean_china > 0.35])/2; positive_china # 174
negative_china <- length(clean_china[clean_china < -0.35])/2; negative_china # 2
total_china <- length(clean_china[clean_china != 0.35])/2; total_china # 35378

write.table(clean_china,
            file = "outputs/r/308_rare_taxa/china/cyt-china-rmt-network.csv",
            sep = "\t", 
            quote = F)
```

## igraph network

```{r}
load("outputs/r/308_rare_taxa/bac-china.RData") # Bacteria_seed_china.p
load("outputs/r/308_rare_taxa/se-china.RData") # se.mb.china_01_10_20
load("outputs/r/308_rare_taxa/betmat-china.RData") # betMat_china

ig.mb.china <- adj2igraph(getRefit(se.mb.china_01_10_20),
                    vertex.attr = list(name = taxa_names(Bacteria_seed_china.p))) 
igraph::transitivity(ig.mb.china) # 0.4048101
china.net <- plot_network(ig.mb.china, Bacteria_seed_china.p, type = 'taxa', color = "Phylum"); china.net
# save(ig.mb.china, file = "outputs/r/308_rare_taxa/ig-mb-china.RData")

otu.ids <- colnames(se.mb.china_01_10_20$data) # spiec.out
edges <- E(ig.mb.china) # spiec.graph
edge.colors <- c()
for(e.index in 1:length(edges)){
	adj.nodes = ends(ig.mb.china, edges[e.index])
	xindex = which(otu.ids == adj.nodes[1])
	yindex = which(otu.ids == adj.nodes[2])
	beta = betaMat_china[xindex, yindex]
	if(beta > 0){
		edge.colors = append(edge.colors, "forestgreen")
	}else if(beta < 0){
		edge.colors = append(edge.colors, "red")
	}
}
E(ig.mb.china)$color <- edge.colors

spiec.graph.b <- ig.mb.china
nodenames <- V(spiec.graph.b)$name
V(spiec.graph.b)$name <- getTaxonomy(nodenames, Bacteria_seed_china.p, useRownames = T)
E(spiec.graph.b)$arrow.size = 5
V(spiec.graph.b)$color = "white"
V(spiec.graph.b)$frame.color = "black"
tkplot(spiec.graph.b)


clusters_china <- cluster_fast_greedy(ig.mb.china)
# clusterOneIndices <- which(clusters$membership == 1)
# clusterOneOtus <- clusters$names[clusterOneIndices]
# clusterTwoIndices <- which(clusters$membership == 2)
# clusterTwoOtus <- clusters$names[clusterTwoIndices]


load("outputs/r/308_rare_taxa/bac-hgr.RData") # Bacteria_seed_hgr.p
load("outputs/r/308_rare_taxa/se-hgr.RData") # se.mb.hgr_01_10_20

ig.mb.hgr <- adj2igraph(getRefit(se.mb.hgr_01_10_20),
                    vertex.attr = list(name = taxa_names(Bacteria_seed_hgr.p)))
igraph::transitivity(ig.mb.hgr) # 0.2917342
hgr.net <- plot_network(ig.mb.hgr, Bacteria_seed_hgr.p, type = 'taxa', color = "Phylum"); hgr.net
# save(ig.mb.hgr, file = "outputs/r/308_rare_taxa/ig-mb-hgr.RData")


load("outputs/r/308_rare_taxa/bac-masterk.RData") # Bacteria_seed_masterk.p
load("outputs/r/308_rare_taxa/se-masterk.RData") # se.mb.masterk_01_10_20

ig.mb.masterk <- adj2igraph(getRefit(se.mb.masterk_01_10_20),
                    vertex.attr = list(name = taxa_names(Bacteria_seed_masterk.p)))
igraph::transitivity(ig.mb.masterk) # 0.3533619
masterk.net <- plot_network(ig.mb.masterk, Bacteria_seed_masterk.p, type = 'taxa', color = "Phylum"); masterk.net
# save(ig.mb.masterk, file = "outputs/r/308_rare_taxa/ig-mb-masterk.RData")

# ggsave(masterk.net, 
#       width = 50,
#       height = 50,
#       units = "cm",
#       scale = 1,
#       file = "outputs/visual/r/308_rare_taxa/igraph-masterk.png")
```

#### Combine igraph networks

```{r}
## set size of vertex proportional to clr-mean
vsize <- rowMeans(clr(Bacteria_seed_china.p, 1)) + 6
am.coord.china <- layout.fruchterman.reingold(ig.mb.china)
vsize <- rowMeans(clr(Bacteria_seed_china.p, 1)) + 6
am.coord.hgr <- layout.fruchterman.reingold(ig.mb.hgr)
vsize <- rowMeans(clr(Bacteria_seed_china.p, 1)) + 6
am.coord.masterk <- layout.fruchterman.reingold(ig.mb.masterk)

par(mfrow = c(1, 3))
plot(ig.mb.china, layout = am.coord.china, vertex.size = vsize, vertex.label = NA, main = "MB")
plot(ig.mb.hgr, layout = am.coord.hgr, vertex.size = vsize, vertex.label = NA, main = "MB")
plot(ig.mb.masterk, layout = am.coord.masterk, vertex.size = vsize, vertex.label = NA, main = "MB")
```

### Import to Cytoscape

```{r}
load("outputs/r/308_rare_taxa/prev-rel-taxo-obs-cultivar-cr.RData") # ASV_prev_rel_taxo_obs_cultivar_cr

# China
load("outputs/r/308_rare_taxa/bac-china.RData") # Bacteria_seed_china.p
load("outputs/r/308_rare_taxa/ig-mb-china.RData") # ig.mb.china

china_taxa <- data.frame(otu_table(Bacteria_seed_china.p)) %>%
  rownames_to_column("ASV")
china_taxa_mer <- inner_join(china_taxa, 
                             ASV_prev_rel_taxo_obs_cultivar_cr, 
                             by = "ASV")

write.table(china_taxa_mer,
            file = "outputs/r/308_rare_taxa/china/cyt-china-asv.csv",
            sep = "\t", 
            quote = F)

write.graph(ig.mb.china,
            file = "outputs/r/308_rare_taxa/china/cyt-china-network.csv",
            format = "ncol")

createNetworkFromIgraph(ig.mb.china, "China_Igraph")

# HGR
load("outputs/r/308_rare_taxa/bac-hgr.RData") # Bacteria_seed_hgr.p
load("outputs/r/308_rare_taxa/ig-mb-hgr.RData") # ig.mb.hgr

hgr_taxa <- data.frame(otu_table(Bacteria_seed_hgr.p)) %>% 
  rownames_to_column("ASV")
hgr_taxa_mer <- inner_join(hgr_taxa, 
                           ASV_prev_rel_taxo_obs_cultivar_cr, 
                           by = "ASV")

write.table(hgr_taxa_mer,
            file = "outputs/r/308_rare_taxa/hgr/cyt-hgr-asv.csv",
            sep = "\t", 
            quote = F)

write.graph(ig.mb.hgr,
            file = "outputs/r/308_rare_taxa/hgr/cyt-hgr-network.csv",
            format = "ncol")

createNetworkFromIgraph(ig.mb.hgr, "HGR_myIgraph")

# Master Kush
load("outputs/r/308_rare_taxa/bac-masterk.RData") # Bacteria_seed_masterk.p
load("outputs/r/308_rare_taxa/ig-mb-masterk.RData") # ig.mb.masterk

masterk_taxa <- data.frame(otu_table(Bacteria_seed_masterk.p)) %>% 
  rownames_to_column("ASV")
masterk_taxa_mer <- inner_join(masterk_taxa, 
                           ASV_prev_rel_taxo_obs_cultivar_cr, 
                           by = "ASV")

write.table(masterk_taxa_mer,
            file = "outputs/r/308_rare_taxa/masterk/cyt-masterk-asv.csv",
            sep = "\t", 
            quote = F)

write.graph(ig.mb.masterk,
            file = "outputs/r/308_rare_taxa/masterk/cyt-masterk-network.csv",
            format = "ncol")

createNetworkFromIgraph(ig.mb.masterk, "MasterK_Igraph")
```

## Network centralities

```{r}
# China
cent_china <- read.table(
  "outputs/cytoscape/cyt-china-network.csv",
  sep = ",",
  header = TRUE,
  fill = TRUE,
  na.strings = "")

asv <- data.frame(otu_table(Bacteria_seed_china))
ASV_rel <- apply(vegan::decostand(asv, 
                                  method = "total", 
                                  MARGIN = 2), 1, mean)

ASV_rel <- rownames_to_column(as.data.frame(ASV_rel),
                              "name") %>% 
  mutate(ASV_rel_p = ASV_rel * 100)

cent_china_ra <- cent_china %>% 
  inner_join(ASV_rel, 
             by = "name") 

# HGR
cent_hgr <- read.table(
  "outputs/cytoscape/cyt-hgr-network.csv",
  sep = ",",
  header = TRUE,
  fill = TRUE,
  na.strings = "")

asv <- data.frame(otu_table(Bacteria_seed_hgr))
ASV_rel <- apply(vegan::decostand(asv, 
                                  method = "total", 
                                  MARGIN = 2), 1, mean)

ASV_rel <- rownames_to_column(as.data.frame(ASV_rel),
                              "name") %>% 
  mutate(ASV_rel_p = ASV_rel * 100)

cent_hgr_ra <- cent_hgr %>% 
  inner_join(ASV_rel, 
             by = "name") 

# Master Kush
cent_mk <- read.table(
  "outputs/cytoscape/cyt-masterk-network.csv",
  sep = ",",
  header = TRUE,
  fill = TRUE,
  na.strings = "")

asv <- data.frame(otu_table(Bacteria_seed_masterk))
ASV_rel <- apply(vegan::decostand(asv, 
                                  method = "total", 
                                  MARGIN = 2), 1, mean)

ASV_rel <- rownames_to_column(as.data.frame(ASV_rel),
                              "name") %>% 
  mutate(ASV_rel_p = ASV_rel * 100)

cent_mk_ra <- cent_mk %>% 
  inner_join(ASV_rel, 
             by = "name") 
```

### Plot abundance and degree

```{r}
color_cr = c(Core = "#C7522B", Rare = "#C7D6B2", Others = "#D3D3D3")

ab_deg <- 
  ggplot(
    cent_hgr_ra,
    aes(x = ASV_rel_p, 
        y = Degree, 
        color = classification)) +
      geom_point(size = 3, 
                 alpha = 0.7) + 
  ggrepel::geom_text_repel(data = subset(cent_hgr_ra,
                                         Degree >= 6),
            aes(ASV_rel_p,
                Degree,
                label = name),
            size = 2,
            show.legend = F) +
  geom_hline(yintercept = 6,
             linetype = "dotted", 
             color = "black") +
  geom_vline(xintercept = 0.1,
             linetype = "dotted",
             color = "black") +
      theme_classic() +
      theme(
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = "right",
        legend.direction = "vertical",
        panel.spacing = unit(0.1, "lines")) +
      scale_x_log10(breaks = c(10, 0.1, 0.001, 0.00001),
                    labels = c(10, 0.1, 0.001, 0.00001)) +
      scale_color_manual(values = color_cr)	+
      labs(x = "Relative Abundance (%)", 
           y = "Degree", 
           color = "")
ab_deg

ggsave("outputs/visual/r/308_rare_taxa/deg-ra-hgr.png",
        width = 20,
        height = 18,
        units = "cm",
        scale = 1,
        dpi = 600)
```

### Plot Betweenness Centrality and degree

```{r}
color_cr = c(Core = "#C7522B", Rare = "#C7D6B2", Others = "#D3D3D3")

bc_deg <- 
  ggplot(
    cent_mk_ra,
    aes(x = BetweennessCentrality, 
        y = Degree, 
        color = classification)) +
      geom_point(size = 3, 
                 alpha = 0.7) + 
  ggrepel::geom_text_repel(data = subset(cent_mk_ra,
                                         BetweennessCentrality >= 0.05 & Degree >= 10),
            aes(BetweennessCentrality,
                Degree,
                label = name),
            size = 2,
            show.legend = F) +
  geom_hline(yintercept = 10,
             linetype = "dotted", 
             color = "black") +
  geom_vline(xintercept = 0.05,
             linetype = "dotted",
             color = "black") +
      theme_classic() +
      theme(
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = "right",
        legend.direction = "vertical",
        panel.spacing = unit(0.1, "lines")) +
      scale_x_log10(breaks = c(1, 0.5, 0.05, 0.005, 0.00005, 0),
                    labels = c(1, 0.5, 0.05, 0.005, 0.00005, 0)) +
      scale_color_manual(values = color_cr)	+
      labs(x = "Betweenness Centrality", 
           y = "Degree", 
           color = "")
bc_deg

ggsave("outputs/visual/r/308_rare_taxa/deg-bc-mk.png",
        width = 20,
        height = 18,
        units = "cm",
        scale = 1,
        dpi = 600)
```

### Plot Closeness Centrality and degree

```{r}
color_cr = c(Core = "#C7522B", Rare = "#C7D6B2", Others = "#D3D3D3")

cc_deg <- 
  ggplot(
    cent_mk_ra,
    aes(x = ClosenessCentrality, 
        y = Degree, 
        color = classification)) +
      geom_point(size = 3, 
                 alpha = 0.7) + 
  ggrepel::geom_text_repel(data = subset(cent_mk_ra,
                         classification != "Core" & ClosenessCentrality >= 0.1 & Degree >= 10),
            aes(ClosenessCentrality,
                Degree,
                label = name),
            size = 2,
            show.legend = F) +
  geom_hline(yintercept = 10,
             linetype = "dotted", 
             color = "black") +
  geom_vline(xintercept = 0.1,
             linetype = "dotted",
             color = "black") +
      theme_classic() +
      theme(
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        legend.text = element_text(size = 10),
        legend.position = "right",
        legend.direction = "vertical",
        panel.spacing = unit(0.1, "lines")) +
      scale_x_log10(breaks = c(1, 0.5, 0.1, 0.01),
                    labels = c(1, 0.5, 0.1, 0.01)) +
      scale_color_manual(values = color_cr)	+
      labs(x = "Closeness Centrality", 
           y = "Degree", 
           color = "")
cc_deg

ggsave("outputs/visual/r/308_rare_taxa/deg-cc-mk.png",
        width = 20,
        height = 18,
        units = "cm",
        scale = 1,
        dpi = 600)
```

<https://ramellose.github.io/networktutorials/workshop_MDA.html>

[***https://kateto.net/netscix2016.html***](https://kateto.net/netscix2016.html){.uri} <https://www.frontiersin.org/articles/10.3389/fcell.2021.719072/full> hubs: <https://cran.r-project.org/web/packages/bipartite/bipartite.pdf> (czvalues) <https://www.mdpi.com/1422-0067/20/9/2130> (RMT) <http://ieg4.rccc.ou.edu/mena/> <https://link.springer.com/article/10.1186/1471-2105-13-113> <https://www.sciencedirect.com/science/article/pii/S0375960106006530> <https://journals.aps.org/pre/pdf/10.1103/PhysRevE.76.026109>

```{r}
# Degree centrality: where the degree is the number of connections a node has
## Scale-free networks have a degree distribution that follows a power law.
## Scale-freeness decreases the network’s sensitivity to random attacks
china.spiec.deg <- degree(ig.mb.china)
hist(china.spiec.deg)
fit <- fit_power_law(china.spiec.deg); fit # p-value = 1, cannot assume that the degree distribution is drawn from a different distribution than the power-law distribution = scale free

# Betweenness centrality: number of shortest paths passing through a node, where the shortest paths are from all nodes to all other nodes
## If a node has low degree but connects two (or more) subcomponents of a network, it can have low degree but high betweenness centrality
china.spiec.bet <- betweenness(ig.mb.china)
hist(china.spiec.bet)

## This measure takes the set of neighbours of a node and uses those to calculate the node’s eigenvector.
china.spiec.eigen <- eigen_centrality(ig.mb.china)
hist(china.spiec.eigen$vector)

# Closeness centrality: Counts the shortest paths from a node to all other nodes. Therefore, it measures how distant a node is to other nodes. 
china.spiec.close <- closeness(ig.mb.china)
china.spiec.close[china.spiec.close < 0.0001] = NA # remove small values caused by disconnected components to better visualize differences in the network
hist(china.spiec.close)
# To plot:
# centralities$closeness <- china.spiec.close
# centralities$closeness[china.spiec.close < 0.0001] = NA

plot_network(ig.mb.china, 
             Bacteria_seed_china.p, 
             type = 'taxa',
             color = "Phylum", 
             label = "Genus") + 
  geom_point(aes(size = china.spiec.bet), # china.spiec.deg, china.spiec.bet, china.spiec.close, china.spiec.eigen$vector
#             colour = "", 
             shape = 1)

# Nodes can have high values of a specific centrality without having high degree.
centralities <- data.frame(degree = china.spiec.deg, betweenness = china.spiec.bet)
ggplot(data = centralities, aes(x = degree, y = betweenness)) +
  geom_point() +
  geom_smooth(method = "auto")

# Other measures:
## HITS algorithm: can datect taxa that have a large effect on other taxa, whereas others taxa are affected by many taxa.
## Freilich and colleagues (2018) co-occurrence networks may not reflect ecological networks, but the effect of environmental conditions. The species that had the strongest correlations (e.g. hub species) were generalist species, not validated keystones.


## Assortativity of a network quantifies whether nodes are more likely to connect to similar nodes or to dissimilar nodes. This can be similarity in degree, edge weight or any other value of interest. Try calculating if nodes in this network preferentially connect to nodes with a similar number of positive or negative edges, or with a similar taxonomy. Is this a disassortative or assortative network?
## The degree assortativity close to zero indicates that nodes with a high degree are slightly more likely to connect to other high-degree nodes. For edge weight assortativity, nodes with a certain number of positive edges are slightly more likely to connect to nodes with a similar number of edges. This is not the case for negative edges. 
assortativity.degree(ig.mb.china) # 0.2946537

# edge weight assortativity
negedges = rep(0, length(V(spiec.graph)))
posedges = rep(0, length(V(spiec.graph)))
connecting_edges <- incident_edges(spiec.graph, V(spiec.graph), mode="all")
i = 1
for (node in connecting_edges){
  if (length(node) != 0){
    for (edge in node){
      weight <- get.edge.attribute(spiec.graph, 'edge_weight')[edge]
      if (weight > 0){
        posedges[i] = posedges[i] + 1
      }
      else if (weight < 0){
        negedges[i] = negedges[i] + 1
      }
    }
  }
  i = i + 1
}
assortativity(spiec.graph, posedges)

assortativity(spiec.graph, negedges)

# taxonomic assortativity
levels <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
for (k in 2:7){
  tax <- tax_table(phyloseqobj.f)[,k]
  # need to convert tax to vertex types; nominal
  tax_dict <- data.frame(unique(tax), seq(1:length(unique(tax))))
  tax_nom <- c()
  for (i in 1:length(tax)){
    tax_nom[[i]] <- tax_dict[tax_dict[,1] == tax[[i]], 2]
  }
  assort <- assortativity_nominal(spiec.graph, tax_nom)
  print(paste(levels[k], assort))
}

# Network permutation is not that commonly adopted for microbial association networks, it can help to identify nodes not robust to errors.
# Carry out a t-test and assess whether the centralities from the permutations are consistently worse than the centrality values measured for the original graph (does not work for eigenvectors).
permlist <- list()
for (i in 1:100){
  permlist[[i]] <- rewire(ig.mb.china, 
                          with = keeping_degseq(niter = 38)) # function to rewire networks: rewire 20% of the edges while preserving the original degree distribution
}

# Instead of comparing centrality values for the separate networks, we are going to test if network centrality values for the top 1, top 5, top 10 and top 10% (19) nodes are still in the top values for the permuted networks. Try to figure this out yourself and see what effect the top # has on robustness.

# The code snippet below checks for each top # whether nodes that were in the original top # are still present in the permuted top #. The ratio of permutations where the nodes kept their positions is the robustness score. Therefore, a score of 1 indicates that nodes are very robust to errors. As you can see in the final figure, the degree robustness is stable at 1. However, all other centralities do not consistently reach 1; nodes shift in ranking more easily, which explains why the top 1 closeness node has lower robustness than most of the top 10 closeness nodes. At the same time, the graph also indicates that some high-ranking nodes have low robustness and may be dependent on one or two specific edges. For example, the eigenvector centrality is quite robust for some nodes, but highly variable for two others.
permresults = list()
for (i in 1:100){
  perm.deg <- degree(permlist[[i]])
  perm.bet <- betweenness(permlist[[i]])
  perm.close <- closeness(permlist[[i]])
  perm.eigen <- eigen_centrality(permlist[[i]])
  perms <- data.frame(degree=perm.deg, betweenness=perm.bet, closeness=perm.close, eigenvector=perm.eigen$vector)
  permresults[[i]] <- perms
}

nums <- c(1, 5, 10, 19)
results <- list()
for (m in 1:4){
  num = nums[[m]]
  data <- data.frame(matrix(0, ncol=4, nrow=4*num))
  colnames(data) <- c('ID', 'centrality', 'robustness', 'top')
  data$top <- num
  topdegree <- rownames(centralities[order(centralities$degree, decreasing=TRUE),])[1:num]
  topbetweenness <- rownames(centralities[order(centralities$betweenness, decreasing=TRUE),])[1:num]
  topcloseness <- rownames(centralities[order(centralities$closeness, decreasing=TRUE),])[1:num]
  topeigen <- rownames(centralities[order(centralities$eigen, decreasing=TRUE),])[1:num]
  data$ID <- c(topdegree, topbetweenness, topcloseness, topeigen)
  data$centrality <- c(rep("degree", num), rep("betweenness", num), rep("closeness", num), rep("eigen", num))
  for (i in 1:100){
    permdegree <- rownames(permresults[[i]][order(permresults[[i]]$degree, decreasing=TRUE),])[1:num]
    permbetweenness <- rownames(permresults[[i]][order(permresults[[i]]$betweenness, decreasing=TRUE),])[1:num]
    permcloseness <- rownames(permresults[[i]][order(permresults[[i]]$closeness, decreasing=TRUE),])[1:num]
    permeigen <- rownames(permresults[[i]][order(permresults[[i]]$eigen, decreasing=TRUE),])[1:num]
    hits <- c(topdegree %in% permdegree, topbetweenness %in% permbetweenness, topcloseness %in% permcloseness, topeigen %in% permeigen)
    data[hits,3] <- data[hits,3] + 1
  }
  data$robustness <- data$robustness / 100
  results[[m]] <- data
}

data <- rbind(results[[1]], results[[2]], results[[3]], results[[4]])

ggplot(data=data, aes(x=data$top, y=data$robustness, color=data$centrality)) + geom_smooth() + geom_point() + 
  labs(x='Number of top nodes', y='Robustness', color='Centrality measure') + scale_y_continuous(limits=c(0,1))


```

## ggnet network: China

```{r}
load("outputs/r/308_rare_taxa/bac-china.RData") # Bacteria_seed_china.p

mycolors <- colorRampPalette(brewer.pal(20, "Paired"))(20)

head(tax_table(Bacteria_seed_china.p))
colnames(tax_table(Bacteria_seed_china.p))
OTU_china <- t(otu_table(Bacteria_seed_china.p))


n.china <- symBeta(getOptBeta(se.mb.china_01_10_20))
colnames(n.china) <- rownames(n.china) <- colnames(OTU_china)

# add log abundance as properties of vertexes/nodes
vsize_china <- log2(apply(OTU_china, 2, mean)) 

nw_china <- graph.adjacency(n.china,
                            mode = 'undirected',
                            add.rownames = T,
                            weighted = T)


genus_china <- map_levels(colnames(OTU_china), from = "best_hit", to = "Genus", tax_table(Bacteria_seed_china.p))
family_china <- map_levels(colnames(OTU_china), from = "best_hit", to = "Family", tax_table(Bacteria_seed_china.p))
phylum_china <- map_levels(colnames(OTU_china), from = "best_hit", to = "Phylum", tax_table(Bacteria_seed_china.p))


E(nw_china)$color <- as.character(cut(E(nw_china)$weight, 
                                      breaks = c( -1, 0, 1), 
                                      labels = c("tomato",  "lightgreen")))

nw_china.net <- intergraph::asNetwork(nw_china)

nw_china.net %v% "nodesize" <- vsize_china
nw_china.net %v% "Genus" <- genus_china
nw_china.net %v% "Family" <- family_china
nw_china.net %v% "Phylum" <- phylum_china

# plot
p_china <- ggnet2(nw_china.net, 
                  node.color = "Phylum", 
                  palette = mycolors,
                  label = T, 
                  node.size = "nodesize",
                  label.size = 2, 
                  edge.color = "color") +
  guides(color = guide_legend(title = "Phylum"),
         size = F); 
p_china

hub_china <- hub_score(nw_china, 
                       scale = T,
                       weights = NULL, 
                       options = arpack_defaults)
hub_china <- hub_china$vector
max(hub_china)
tail(sort(hub_china), 10)
aut_china <- authority_score(nw_china, weights = NA)$vector
tail(sort(aut_china), 5)
```

### Import to Cytoscape

```{r}
# The resulting edge list needs to be edited to add a header line (one header for each column) and to replace white spaces by commas. 
# This modified edge list file can then be directly imported into Cytoscape via File -> Import -> Network -> File. 

write.graph(spiec.graph,
            file = file.path(output.path,"spieceasi.ncol.txt"), 
            format = "ncol")
```

## ggnet network: Hohenthurmer G. R.

```{r}
load("outputs/r/308_rare_taxa/bac-hgr.RData") # Bacteria_seed_hgr.p

mycolors <- colorRampPalette(brewer.pal(20, "Paired"))(20)

head(tax_table(Bacteria_seed_hgr.p))
colnames(tax_table(Bacteria_seed_hgr.p))
OTU_hgr <- t(otu_table(Bacteria_seed_hgr.p))


n.hgr <- symBeta(getOptBeta(se.mb.hgr_01_10_20))
colnames(n.hgr) <- rownames(n.hgr) <- colnames(OTU_hgr)

# add log abundance as properties of vertexes/nodes
vsize_hgr <- log2(apply(OTU_hgr, 2, mean)) 

nw_hgr <- graph.adjacency(n.hgr,
                            mode = 'undirected',
                            add.rownames = T,
                            weighted = T)


genus_hgr <- map_levels(colnames(OTU_hgr), from = "best_hit", to = "Genus", tax_table(Bacteria_seed_hgr.p))
family_hgr <- map_levels(colnames(OTU_hgr), from = "best_hit", to = "Family", tax_table(Bacteria_seed_hgr.p))
phylum_hgr <- map_levels(colnames(OTU_hgr), from = "best_hit", to = "Phylum", tax_table(Bacteria_seed_hgr.p))

E(nw_hgr)$color <- as.character(cut(E(nw_hgr)$weight, 
                                      breaks = c( -1, 0, 1), 
                                      labels = c("tomato",  "lightgreen")))

nw_hgr.net <- intergraph::asNetwork(nw_hgr)

nw_hgr.net %v% "nodesize" <- vsize_hgr
nw_hgr.net %v% "Genus" <- genus_hgr
nw_hgr.net %v% "Family" <- family_hgr
nw_hgr.net %v% "Phylum" <- phylum_hgr

# plot
p_hgr <- ggnet2(nw_hgr.net, 
                  node.color = "Phylum", 
                  palette = mycolors,
                  label = T, 
                  node.size = "nodesize",
                  label.size = 2, 
                  edge.color = "color") +
  guides(color = guide_legend(title = "Phylum"),
         size = F); 
p_hgr

hub_hgr <- hub_score(nw_hgr, 
                       scale = T,
                       weights = NULL, 
                       options = arpack_defaults)
hub_hgr <- hub_hgr$vector
max(hub_hgr)
tail(sort(hub_hgr), 10)
aut_hgr <- authority_score(nw_hgr, weights = NA)$vector
tail(sort(aut_hgr), 5)
```

### Import to Cytoscape

```{r}
# The resulting edge list needs to be edited to add a header line (one header for each column) and to replace white spaces by commas. 
# This modified edge list file can then be directly imported into Cytoscape via File -> Import -> Network -> File. 

write.graph(spiec.graph,
            file = file.path(output.path,"spieceasi.ncol.txt"), 
            format = "ncol")
```

## ggnet network: Master Kush

```{r}
load("outputs/r/308_rare_taxa/bac-masterk.RData") # Bacteria_seed_masterk.p

mycolors <- colorRampPalette(brewer.pal(20, "Paired"))(20)

head(tax_table(Bacteria_seed_masterk.p))
colnames(tax_table(Bacteria_seed_masterk.p))
OTU_masterk <- t(otu_table(Bacteria_seed_masterk.p))

n.masterk <- symBeta(getOptBeta(se.mb.masterk_01_10_20))
colnames(n.masterk) <- rownames(n.masterk) <- colnames(OTU_masterk)

# add log abundance as properties of vertexes/nodes
vsize_masterk <- log2(apply(OTU_masterk, 2, mean)) 

vcolor_masterk <- 

nw_masterk <- graph.adjacency(n.masterk,
                            mode = 'undirected',
                            add.rownames = T,
                            weighted = T)


genus_masterk <- map_levels(colnames(OTU_masterk), from = "best_hit", to = "Genus", data = tax_table(Bacteria_seed_masterk.p))
family_masterk <- map_levels(colnames(OTU_masterk), from = "best_hit", to = "Family", data = tax_table(Bacteria_seed_masterk.p))
phylum_masterk <- map_levels(colnames(OTU_masterk), from = "best_hit", to = "Phylum", data = tax_table(Bacteria_seed_masterk.p))


E(nw_masterk)$color <- as.character(cut(E(nw_masterk)$weight, 
                                      breaks = c( -1, 0, 1), 
                                      labels = c("tomato",  "lightgreen")))

nw_masterk.net <- intergraph::asNetwork(nw_masterk)

nw_masterk.net %v% "nodesize" <- vsize_masterk
nw_masterk.net %v% "Genus" <- genus_masterk
nw_masterk.net %v% "Family" <- family_masterk
nw_masterk.net %v% "Phylum" <- phylum_masterk

# plot
p_masterk <- ggnet2(nw_masterk.net, 
                  node.color = "Phylum", 
                  palette = mycolors,
                  label = T, 
                  node.size = "nodesize",
                  label.size = 2, 
                  edge.color = "color") +
  guides(color = guide_legend(title = "Phylum"),
         size = F); 
p_masterk

hub_masterk <- hub_score(nw_masterk, 
                       scale = T,
                       weights = NULL, 
                       options = arpack_defaults)
hub_masterk <- hub_masterk$vector
max(hub_masterk)
tail(sort(hub_masterk), 10)
aut_masterk <- authority_score(nw_masterk, weights = NA)$vector
tail(sort(aut_masterk), 5)
```

### Import to Cytoscape

```{r}
# The resulting edge list needs to be edited to add a header line (one header for each column) and to replace white spaces by commas. 
# This modified edge list file can then be directly imported into Cytoscape via File -> Import -> Network -> File. 

write.graph(spiec.graph,
            file = file.path(output.path,"spieceasi.ncol.txt"), 
            format = "ncol")
```

#### Combine ggnet networks

```{r}
combined_networks <- ggpubr::ggarrange(p_china, p_hgr, p_masterk,
                                       nrow = 1,
                                       ncol = 3,
                                       labels = "AUTO", 
                                       common.legend = TRUE); combined_networks
```
