---
title: "Functional Prediction"
author: "Carolina Lobato"
date: "2023-11-07"
output: html_document
---

# Load libraries

```{r, echo = FALSE}
library(phyloseq)
library(microbiome)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(colorspace)
library(RColorBrewer)
library(ggpubr)
# install.packages("zCompositions")
library(zCompositions)
```

# Prepare merged table for running faprotax in python 3

```{r}
load("outputs/r/204_setup/bh.RData")

# OTU table
asv <- data.frame(otu_table(bh)); asv[1:4,1:4]

# Calculate the total read count for your samples to use later
raw.depth <- colSums(asv)
names(raw.depth) <- colnames(asv)
raw.depth

# Tax info
taxdf <- data.frame(tax_table(bh)); taxdf[1:4,1:4]
taxdf$taxstring = paste(taxdf$Kingdom, taxdf$Phylum, taxdf$Class, taxdf$Order, taxdf$Family, taxdf$Genus, sep="; ")
taxdf[1:4,]

# Append the taxonomy string as a column to the otu table. 
asv$taxonomy = taxdf[as.vector(row.names(asv)),"taxstring"]
asv[1:4,451:452]

write.table(asv, 
            "outputs/faprotax/faprotax.taxon.table.tsv", 
            sep = "\t", 
            row.names = T,
            quote = F)
```

# Import results from Faprotax

```{r}
fun <- read.table("outputs/faprotax/out.functable.tsv", 
                  sep = "\t", 
                  header = T,
                  row.names = 1)

# Discard functions identified in fewer than 10 reads
fun <- fun[which(rowSums(fun) > 10),]
# Discard samples where no functions were identified
fun <- fun[,which(colSums(fun) > 1)]
fun[1:4,1:4]
dim(fun) # 46 different functions id
```

# Normalize the counts to the total function identifications for each sample

```{r}
rel.fun <- apply(fun,
                 MARGIN = 2, 
                 FUN = function(x) x/sum(x))

rel.fun <- data.frame(rel.fun); rel.fun
```

# Add metadata

```{r}
# Load metadata
meta <- meta(bh); meta

# Melt fun table 
rel.fun$BioFunction = row.names(rel.fun)
melt.fun <- reshape::melt(rel.fun, id.vars = "BioFunction")

# Make this a sparse dataframe
melt.s.fun <- melt.fun[which(melt.fun$value > 0),] 

# Pull relevant metadata and add to the melted dataframe.
colnames(melt.s.fun) <- c("BioFunction", "sample", "RA")
melt.s.fun$cultivar <- meta[as.vector(melt.s.fun$sample),"cultivar"]
melt.s.fun$domestication <- meta[as.vector(melt.s.fun$sample),"domestication"]
melt.s.fun$chemotype <- meta[as.vector(melt.s.fun$sample),"chemotype"]
melt.s.fun$inflorescence <- meta[as.vector(melt.s.fun$sample),"inflorescence"]

melt.s.fun[1:10,]
```

# Plot: how frequently OTUs were assigned a function out of all possible assignments

```{r}
# Inappropriate for this kind of data
mycolors <- colorRampPalette(brewer.pal(12, "Paired"))(12)
theme_set(theme_classic(base_size = 12))

ggplot(melt.s.fun[which(melt.s.fun$RA >= 0.1),], 
       aes(x = cultivar,
           y = RA,
           fill = BioFunction)) +
  geom_bar(stat = "identity", 
           position = "fill") + 
  theme(
    axis.text.x = (element_text(angle = 90,
                                hjust = 1,
                                vjust = 1,
                                size = 8)),
    legend.position = "right",
    legend.box = "vertical",
    legend.direction = "vertical",
    legend.key.size = unit(0.3, 'cm'),
    legend.text = element_text(size = 6),
    strip.background = element_blank(),
    strip.text = element_blank()) +
  scale_fill_manual(values = c(mycolors, "pink")) +
  labs(x = "",
       y = "",
       fill = "Biofunctions")
```

# Relative Scaling

```{r}
rel.fun2 <- sapply(colnames(fun), 
                   function(x) fun[,x]/raw.depth[x])

rel.fun2 <- data.frame(rel.fun2)
row.names(rel.fun2) <- row.names(rel.fun2)

rel.fun2[1:10,1:4]

rel.fun2$BioFunction = row.names(rel.fun2)
melt.fun2 <- reshape::melt(rel.fun2, id.vars = "BioFunction")

# Make this a sparse dataframe
melt.s.fun2 <- melt.fun2[which(melt.fun2$value > 0),] 

# Pull relevant metadata and add to the melted dataframe.
colnames(melt.s.fun2) <- c("BioFunction", "sample", "RS")
melt.s.fun2$cultivar <- meta[as.vector(melt.s.fun2$sample),"cultivar"]
melt.s.fun2$domestication <- meta[as.vector(melt.s.fun2$sample),"domestication"]
melt.s.fun2$chemotype <- meta[as.vector(melt.s.fun2$sample),"chemotype"]
melt.s.fun2$inflorescence <- meta[as.vector(melt.s.fun2$sample),"inflorescence"]

melt.s.fun2[1:10,]
```

# Heat maps: They don't imply a composition

```{r}
melt.s.fun2$BioFunction <- as.factor(melt.s.fun2$BioFunction)
melt.s.fun2$ShortFunction <- melt.s.fun2$BioFunction
levels(melt.s.fun2$ShortFunction) <- c("CH4_CO2", "CH4_H", "CH4", "CH3OH_OXID", "METHYLOT", "SO4_RESP", "S_RESP", "NO3_DENITR", "NO2_DENITR", "N2O_DENITR", "DENITR", "CHIT", "H_OXID", "N2_FIX", "NO2_RESP", "CELLU", "XYTAN", "MN_OXID", "FERM", "AERO_CH_HETR", "IN_PAR", "HU_PATH", "HU_GU", "HU", "MAM_GU", "AN_PS", "AROM_HC_DEG", "AROM_DEG", "ALIPH_NMHC_DEG", "HC_DEG", "FE_OXI", "NO3_RESP", "NO3_RED", "N_RESP", "IC_P", "PRED_EXOP", "NPHY_CYAN", "PHY_CYAN", "ANOXY_PHAT_SOXY", "ANOXY_PHAT", "OXY_PHAT", "PHAT", "PHHT", "PHT", "UREO", "CHHT")
sorted.melt <- melt.s.fun2

ggplot(sorted.melt[which(sorted.melt$RS >= 0.02),], 
       aes(x = cultivar,
           y = tolower(ShortFunction), 
           fill = RS)) + 
  geom_tile() + 
  scale_fill_viridis_c(limits = c(0, 1)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90,
                                hjust = 1,
                                size = 10)) +
  labs(y = "FAPROTAX Function",
       x = "",
       fill = "Count")

ggsave("outputs/visual/r/309_faprotax/fun-rs-cultivar.png",
       width = 24,
       height = 16,
       units = "cm",
       scale = 1,
       dpi = 600)
       
ggplot(sorted.melt[which(sorted.melt$RS >= 0.02 & !is.na(sorted.melt$chemotype)),], 
       aes(x = chemotype, 
           y = tolower(ShortFunction), 
           fill = RS)) +
  geom_tile() + 
  scale_fill_viridis_c(limits = c(0, 0.45)) + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90,
                                hjust = 1,
                                size = 10)) +
  labs(y = "FAPROTAX Function",
       x = "",
       fill = "Count")

ggsave("outputs/visual/r/309_faprotax/fun-rs-chem.png",
       width = 24,
       height = 16,
       units = "cm",
       scale = 1,
       dpi = 600)
```

# Taxonomy Details

```{r}
n2fix.df <- read.csv("outputs/faprotax/out_subtables/nitrogen_fixation.txt", 
                     sep = "\t", 
                     header = TRUE)
n2fix.df$taxonomy <- row.names(n2fix.df)

# Extract genus names
n2fix.Genus <- as.vector(sapply(n2fix.df$record, 
                                function(x) str_split(x, "; ")[[1]][6])); n2fix.Genus # 49 ASVs

n2.otus <- row.names(taxdf[which(taxdf$Genus %in% n2fix.Genus),]); n2.otus # 49 ASVs

# Relative Abundance
n2.rel <- transform_sample_counts(bh,
                                  function(x) x/sum(x)) %>%
  prune_taxa(taxa_names(bh) %in% n2.otus, .) %>% 
  prune_samples(sample_sums(.) > 0,.)

n2.rel
```

# Proportional Abundance

```{r}
transform.clr <- function(otu.df){
  d.1 <- data.frame(otu.df[which(rowSums(otu.df) > 10),],
                    check.names = F)
  d.czm <- t(zCompositions::cmultRepl(t(d.1), 
                                      label = 0, 
                                      method = "CZM"))
  d.clr <- apply(d.czm, 2, 
                 function(x){log(x) - mean(log(x))})
  return(d.clr)
}
```

```{r}
# Does it matter how abundant the nitrate respiring OTUs are relative to all of the OTUs in a sample, or do you care about their abundance relative only to other nitrate respiring OTUs within a sample? In this case, Iâ€™ve chosen to scale relative to all OTUs.

rel.asv <- transform.clr(data.frame(otu_table(bh)))
rel.bh <- merge_phyloseq(otu_table(rel.asv,
                                   taxa_are_rows = T), 
                         tax_table(bh),
                         sample_data(bh))

n2.bh.rel <- prune_taxa(taxa_names(rel.bh) %in% n2.otus, rel.bh)

n2.melt <- psmelt(n2.bh.rel)

# Make a new variable with OTU and Genus labels
n2.melt$Otu.Genus <- paste(n2.melt$Genus, 
                           n2.melt$OTU, 
                           sep = ": ")


```
